{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-10-22T16:03:19.397848Z",
     "start_time": "2025-10-22T16:03:18.538372Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "GTF_PATH = \"../datasets/gencode.v49.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "miRNA_GTF_PATH = \"../datasets/preprocessed/miRNA_gtf.csv\"\n",
    "\n",
    "pd.set_option('display.max_colwidth', None)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T16:03:19.413824Z",
     "start_time": "2025-10-22T16:03:19.400808Z"
    }
   },
   "id": "3d4200b452b00c40",
   "execution_count": 2
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "(8063229, 9)"
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = [\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "\n",
    "gtf_df = pd.read_csv(GTF_PATH, sep='\\t', comment='#', names=cols)\n",
    "gtf_df.shape"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T16:04:39.732631Z",
     "start_time": "2025-10-22T16:04:09.635310Z"
    }
   },
   "id": "cfded71d977db18c",
   "execution_count": 6
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\giorg\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\numpy\\_core\\fromnumeric.py:57: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.\n",
      "  return bound(*args, **kwds)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chunk 1/10 ...\n",
      "Processing chunk 2/10 ...\n",
      "Processing chunk 3/10 ...\n",
      "Processing chunk 4/10 ...\n",
      "Processing chunk 5/10 ...\n",
      "Processing chunk 6/10 ...\n",
      "Processing chunk 7/10 ...\n",
      "Processing chunk 8/10 ...\n",
      "Processing chunk 9/10 ...\n",
      "Processing chunk 10/10 ...\n"
     ]
    }
   ],
   "source": [
    "def parse_attributes(attr):\n",
    "    d = {}\n",
    "    for a in attr.split(\";\"):\n",
    "        a = a.strip()\n",
    "        if a == \"\":\n",
    "            continue\n",
    "        if \" \" in a:\n",
    "            key, val = a.split(\" \", 1)\n",
    "            d[key] = val.strip('\"')\n",
    "    return d\n",
    "\n",
    "chunks = np.array_split(gtf_df, 10)\n",
    "results = []\n",
    "for i, chunk in enumerate(chunks):\n",
    "    print(f\"Processing chunk {i+1}/{len(chunks)} ...\")    \n",
    "    attr_df = chunk['attribute'].apply(parse_attributes).apply(pd.Series)\n",
    "    merged = pd.concat([chunk.drop(columns=['attribute']), attr_df], axis=1)    \n",
    "    results.append(merged)\n",
    "\n",
    "complete_gtf_df = pd.concat(results, ignore_index=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T16:27:25.080846Z",
     "start_time": "2025-10-22T16:04:39.734572Z"
    }
   },
   "id": "c934878177e258ce",
   "execution_count": 7
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Try to search for a specific miRNA ID"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "99627924a46c083c"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Empty DataFrame\nColumns: [seqname, source, feature, start, end, score, strand, frame, gene_id, gene_type, gene_name, level, tag, transcript_id, transcript_type, transcript_name, exon_number, exon_id, transcript_support_level, havana_transcript, hgnc_id, havana_gene, ont, protein_id, ccdsid, artif_dupl]\nIndex: []\n\n[0 rows x 26 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>gene_type</th>\n      <th>...</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n      <th>transcript_support_level</th>\n      <th>havana_transcript</th>\n      <th>hgnc_id</th>\n      <th>havana_gene</th>\n      <th>ont</th>\n      <th>protein_id</th>\n      <th>ccdsid</th>\n      <th>artif_dupl</th>\n    </tr>\n  </thead>\n  <tbody>\n  </tbody>\n</table>\n<p>0 rows × 26 columns</p>\n</div>"
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = complete_gtf_df[complete_gtf_df.astype(str).apply(lambda x: x.str.contains('hsa-let-7a-1', case=False, na=False)).any(axis=1)]\n",
    "result.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:06:59.871667Z",
     "start_time": "2025-10-22T17:05:10.148835Z"
    }
   },
   "id": "6dc862826c41d6b1",
   "execution_count": 8
  },
  {
   "cell_type": "markdown",
   "source": [
    "No results found about 'hsa-let-7a-1'. Try another miRNA ID."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2f88d2021464634b"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Empty DataFrame\nColumns: [seqname, source, feature, start, end, score, strand, frame, gene_id, gene_type, gene_name, level, tag, transcript_id, transcript_type, transcript_name, exon_number, exon_id, transcript_support_level, havana_transcript, hgnc_id, havana_gene, ont, protein_id, ccdsid, artif_dupl]\nIndex: []\n\n[0 rows x 26 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>gene_type</th>\n      <th>...</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n      <th>transcript_support_level</th>\n      <th>havana_transcript</th>\n      <th>hgnc_id</th>\n      <th>havana_gene</th>\n      <th>ont</th>\n      <th>protein_id</th>\n      <th>ccdsid</th>\n      <th>artif_dupl</th>\n    </tr>\n  </thead>\n  <tbody>\n  </tbody>\n</table>\n<p>0 rows × 26 columns</p>\n</div>"
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = complete_gtf_df[complete_gtf_df.astype(str).apply(lambda x: x.str.contains('hsa-mir-103b-1', case=False, na=False)).any(axis=1)]\n",
    "result.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:09:43.253473Z",
     "start_time": "2025-10-22T17:07:26.766279Z"
    }
   },
   "id": "8d1b3d8548cf83b2",
   "execution_count": 9
  },
  {
   "cell_type": "markdown",
   "source": [
    "No results found about 'hsa-mir-103b-1'."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "41942e37954f2fc2"
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Filter miRNA entries by searching for 'miRNA' keyword"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f07b3f1e9b2e5e52"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "(5835, 26)"
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_df = complete_gtf_df[complete_gtf_df.astype(str).apply(lambda x: x.str.contains('miRNA', case=False, na=False)).any(axis=1)]\n",
    "mirna_df.shape"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T14:39:41.386603Z",
     "start_time": "2025-10-22T14:38:01.902560Z"
    }
   },
   "id": "bd684500b5eedd43",
   "execution_count": 16
  },
  {
   "cell_type": "markdown",
   "source": [
    "It is sufficient to filter only by **'gene_type'** column."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3e9fd71debebf4ea"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "(5835, 26)"
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_df = complete_gtf_df[complete_gtf_df['gene_type'] == 'miRNA']\n",
    "mirna_df.shape"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T15:29:42.297853Z",
     "start_time": "2025-10-22T15:29:29.205733Z"
    }
   },
   "id": "baea6b6c295286",
   "execution_count": 31
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "     seqname   source     feature  start    end score strand frame  \\\n2415    chr1  ENSEMBL        gene  17369  17436     .      -     .   \n2416    chr1  ENSEMBL  transcript  17369  17436     .      -     .   \n2417    chr1  ENSEMBL        exon  17369  17436     .      -     .   \n2433    chr1  ENSEMBL        gene  30366  30503     .      +     .   \n2434    chr1  ENSEMBL  transcript  30366  30503     .      +     .   \n\n                gene_id gene_type  ... exon_number            exon_id  \\\n2415  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n2416  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n2417  ENSG00000278267.1     miRNA  ...           1  ENSE00003746039.1   \n2433  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n2434  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n\n     transcript_support_level havana_transcript     hgnc_id havana_gene  ont  \\\n2415                      NaN               NaN  HGNC:50039         NaN  NaN   \n2416                       NA               NaN  HGNC:50039         NaN  NaN   \n2417                       NA               NaN  HGNC:50039         NaN  NaN   \n2433                      NaN               NaN  HGNC:35294         NaN  NaN   \n2434                       NA               NaN  HGNC:35294         NaN  NaN   \n\n     protein_id ccdsid artif_dupl  \n2415        NaN    NaN        NaN  \n2416        NaN    NaN        NaN  \n2417        NaN    NaN        NaN  \n2433        NaN    NaN        NaN  \n2434        NaN    NaN        NaN  \n\n[5 rows x 26 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>gene_type</th>\n      <th>...</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n      <th>transcript_support_level</th>\n      <th>havana_transcript</th>\n      <th>hgnc_id</th>\n      <th>havana_gene</th>\n      <th>ont</th>\n      <th>protein_id</th>\n      <th>ccdsid</th>\n      <th>artif_dupl</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>2415</th>\n      <td>chr1</td>\n      <td>ENSEMBL</td>\n      <td>gene</td>\n      <td>17369</td>\n      <td>17436</td>\n      <td>.</td>\n      <td>-</td>\n      <td>.</td>\n      <td>ENSG00000278267.1</td>\n      <td>miRNA</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>HGNC:50039</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2416</th>\n      <td>chr1</td>\n      <td>ENSEMBL</td>\n      <td>transcript</td>\n      <td>17369</td>\n      <td>17436</td>\n      <td>.</td>\n      <td>-</td>\n      <td>.</td>\n      <td>ENSG00000278267.1</td>\n      <td>miRNA</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NA</td>\n      <td>NaN</td>\n      <td>HGNC:50039</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2417</th>\n      <td>chr1</td>\n      <td>ENSEMBL</td>\n      <td>exon</td>\n      <td>17369</td>\n      <td>17436</td>\n      <td>.</td>\n      <td>-</td>\n      <td>.</td>\n      <td>ENSG00000278267.1</td>\n      <td>miRNA</td>\n      <td>...</td>\n      <td>1</td>\n      <td>ENSE00003746039.1</td>\n      <td>NA</td>\n      <td>NaN</td>\n      <td>HGNC:50039</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2433</th>\n      <td>chr1</td>\n      <td>ENSEMBL</td>\n      <td>gene</td>\n      <td>30366</td>\n      <td>30503</td>\n      <td>.</td>\n      <td>+</td>\n      <td>.</td>\n      <td>ENSG00000284332.1</td>\n      <td>miRNA</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>HGNC:35294</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2434</th>\n      <td>chr1</td>\n      <td>ENSEMBL</td>\n      <td>transcript</td>\n      <td>30366</td>\n      <td>30503</td>\n      <td>.</td>\n      <td>+</td>\n      <td>.</td>\n      <td>ENSG00000284332.1</td>\n      <td>miRNA</td>\n      <td>...</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NA</td>\n      <td>NaN</td>\n      <td>HGNC:35294</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n<p>5 rows × 26 columns</p>\n</div>"
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirna_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T15:00:20.988955Z",
     "start_time": "2025-10-22T15:00:20.961268Z"
    }
   },
   "id": "e3cd7ad873d19653",
   "execution_count": 24
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Save the preprocessed miRNA GTF data as CSV file"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "b1c9d0a2a0ce9388"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "mirna_df.to_csv(miRNA_GTF_PATH, index=False)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T15:32:02.018850Z",
     "start_time": "2025-10-22T15:32:01.776742Z"
    }
   },
   "id": "7c3b090c640a2529",
   "execution_count": 32
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Reload created CSV file"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2801e51e0912a42a"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "(5835, 26)"
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(miRNA_GTF_PATH)\n",
    "df.shape"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:10:53.898931Z",
     "start_time": "2025-10-22T17:10:53.700548Z"
    }
   },
   "id": "ad6a9732fd68caf9",
   "execution_count": 10
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "              start           end   level  exon_number  \\\ncount  5.835000e+03  5.835000e+03  5835.0       1945.0   \nmean   7.399684e+07  7.399692e+07     3.0          1.0   \nstd    5.639511e+07  5.639511e+07     0.0          0.0   \nmin    2.050000e+03  2.164000e+03     3.0          1.0   \n25%    2.895858e+07  2.895866e+07     3.0          1.0   \n50%    6.119980e+07  6.119988e+07     3.0          1.0   \n75%    1.102849e+08  1.102849e+08     3.0          1.0   \nmax    2.488264e+08  2.488264e+08     3.0          1.0   \n\n       transcript_support_level  havana_transcript  havana_gene  ont  \\\ncount                       0.0                0.0          0.0  0.0   \nmean                        NaN                NaN          NaN  NaN   \nstd                         NaN                NaN          NaN  NaN   \nmin                         NaN                NaN          NaN  NaN   \n25%                         NaN                NaN          NaN  NaN   \n50%                         NaN                NaN          NaN  NaN   \n75%                         NaN                NaN          NaN  NaN   \nmax                         NaN                NaN          NaN  NaN   \n\n       protein_id  ccdsid  artif_dupl  \ncount         0.0     0.0         0.0  \nmean          NaN     NaN         NaN  \nstd           NaN     NaN         NaN  \nmin           NaN     NaN         NaN  \n25%           NaN     NaN         NaN  \n50%           NaN     NaN         NaN  \n75%           NaN     NaN         NaN  \nmax           NaN     NaN         NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>start</th>\n      <th>end</th>\n      <th>level</th>\n      <th>exon_number</th>\n      <th>transcript_support_level</th>\n      <th>havana_transcript</th>\n      <th>havana_gene</th>\n      <th>ont</th>\n      <th>protein_id</th>\n      <th>ccdsid</th>\n      <th>artif_dupl</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>count</th>\n      <td>5.835000e+03</td>\n      <td>5.835000e+03</td>\n      <td>5835.0</td>\n      <td>1945.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n    </tr>\n    <tr>\n      <th>mean</th>\n      <td>7.399684e+07</td>\n      <td>7.399692e+07</td>\n      <td>3.0</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>std</th>\n      <td>5.639511e+07</td>\n      <td>5.639511e+07</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>min</th>\n      <td>2.050000e+03</td>\n      <td>2.164000e+03</td>\n      <td>3.0</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>25%</th>\n      <td>2.895858e+07</td>\n      <td>2.895866e+07</td>\n      <td>3.0</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>50%</th>\n      <td>6.119980e+07</td>\n      <td>6.119988e+07</td>\n      <td>3.0</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>75%</th>\n      <td>1.102849e+08</td>\n      <td>1.102849e+08</td>\n      <td>3.0</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>max</th>\n      <td>2.488264e+08</td>\n      <td>2.488264e+08</td>\n      <td>3.0</td>\n      <td>1.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.set_option('display.max_colwidth', None)\n",
    "df.describe()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:10:54.067443Z",
     "start_time": "2025-10-22T17:10:53.976231Z"
    }
   },
   "id": "f04d975255ee88fe",
   "execution_count": 11
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Empty DataFrame\nColumns: [seqname, source, feature, start, end, score, strand, frame, gene_id, gene_type, gene_name, level, tag, transcript_id, transcript_type, transcript_name, exon_number, exon_id, transcript_support_level, havana_transcript, hgnc_id, havana_gene, ont, protein_id, ccdsid, artif_dupl]\nIndex: []\n\n[0 rows x 26 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>gene_type</th>\n      <th>...</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n      <th>transcript_support_level</th>\n      <th>havana_transcript</th>\n      <th>hgnc_id</th>\n      <th>havana_gene</th>\n      <th>ont</th>\n      <th>protein_id</th>\n      <th>ccdsid</th>\n      <th>artif_dupl</th>\n    </tr>\n  </thead>\n  <tbody>\n  </tbody>\n</table>\n<p>0 rows × 26 columns</p>\n</div>"
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = df[df.astype(str).apply(lambda x: x.str.contains('hsa-let-7a-1', case=False, na=False)).any(axis=1)]\n",
    "result.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:10:54.205647Z",
     "start_time": "2025-10-22T17:10:54.069593Z"
    }
   },
   "id": "76003a224a5790c7",
   "execution_count": 12
  },
  {
   "cell_type": "markdown",
   "source": [
    "No results found about 'hsa-let-7a-1'."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5d91d086ab767879"
  },
  {
   "cell_type": "markdown",
   "source": [
    "#  Homo sapiens GTF"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "63d776355715ba9b"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "(8063229, 9)"
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HOMO_GTF_PATH = \"../datasets/Homo_sapiens.GRCh38.115.gtf\"\n",
    "cols = [\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "\n",
    "homo_df = pd.read_csv(GTF_PATH, sep='\\t', comment='#', names=cols)\n",
    "homo_df.shape"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:12:20.839253Z",
     "start_time": "2025-10-22T17:11:44.271648Z"
    }
   },
   "id": "567365e022c685cf",
   "execution_count": 13
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\users\\giorg\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\numpy\\_core\\fromnumeric.py:57: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.\n",
      "  return bound(*args, **kwds)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing chunk 1/10 ...\n",
      "Processing chunk 2/10 ...\n",
      "Processing chunk 3/10 ...\n",
      "Processing chunk 4/10 ...\n",
      "Processing chunk 5/10 ...\n",
      "Processing chunk 6/10 ...\n",
      "Processing chunk 7/10 ...\n",
      "Processing chunk 8/10 ...\n",
      "Processing chunk 9/10 ...\n",
      "Processing chunk 10/10 ...\n"
     ]
    }
   ],
   "source": [
    "def parse_attributes(attr):\n",
    "    d = {}\n",
    "    for a in attr.split(\";\"):\n",
    "        a = a.strip()\n",
    "        if a == \"\":\n",
    "            continue\n",
    "        if \" \" in a:\n",
    "            key, val = a.split(\" \", 1)\n",
    "            d[key] = val.strip('\"')\n",
    "    return d\n",
    "\n",
    "chunks = np.array_split(homo_df, 10)\n",
    "results = []\n",
    "for i, chunk in enumerate(chunks):\n",
    "    print(f\"Processing chunk {i+1}/{len(chunks)} ...\")    \n",
    "    attr_df = chunk['attribute'].apply(parse_attributes).apply(pd.Series)\n",
    "    merged = pd.concat([chunk.drop(columns=['attribute']), attr_df], axis=1)    \n",
    "    results.append(merged)\n",
    "\n",
    "homo_complete_gtf_df = pd.concat(results, ignore_index=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:29:38.032929Z",
     "start_time": "2025-10-22T17:12:20.853342Z"
    }
   },
   "id": "d0f2657b40e93184",
   "execution_count": 14
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "Empty DataFrame\nColumns: [seqname, source, feature, start, end, score, strand, frame, gene_id, gene_type, gene_name, level, tag, transcript_id, transcript_type, transcript_name, exon_number, exon_id, transcript_support_level, havana_transcript, hgnc_id, havana_gene, ont, protein_id, ccdsid, artif_dupl]\nIndex: []\n\n[0 rows x 26 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>gene_type</th>\n      <th>...</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n      <th>transcript_support_level</th>\n      <th>havana_transcript</th>\n      <th>hgnc_id</th>\n      <th>havana_gene</th>\n      <th>ont</th>\n      <th>protein_id</th>\n      <th>ccdsid</th>\n      <th>artif_dupl</th>\n    </tr>\n  </thead>\n  <tbody>\n  </tbody>\n</table>\n<p>0 rows × 26 columns</p>\n</div>"
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = homo_complete_gtf_df[homo_complete_gtf_df.astype(str).apply(lambda x: x.str.contains('hsa-let-7a-1', case=False, na=False)).any(axis=1)]\n",
    "result.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-10-22T17:32:55.586713Z",
     "start_time": "2025-10-22T17:29:38.095055Z"
    }
   },
   "id": "225a9f7d26bd74b2",
   "execution_count": 15
  },
  {
   "cell_type": "markdown",
   "source": [
    "No results found about 'hsa-let-7a-1'."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "125103dc7eff1f7d"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
