{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bda669cc",
   "metadata": {},
   "source": [
    "# DATA NORMALIZATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b436641",
   "metadata": {},
   "source": [
    "Various factors affect transcript quantification in RNA-seq data, such as sequencing depth, transcript length, and sample-to-sample and batch-to-batch variability. Normalization methods exist to minimize these variables and ensure reliable transcriptomic data. Sequencing technologies introduce technical variability. Therefore, raw transcriptomic data must be adjusted to account for these technical factors. \n",
    "\n",
    "It is essential to choose the correct RNA-seq normalization method for the dataset and there are three main RNA-seq normalization stages to consider:\n",
    " 1. **Within sample**: Within sample normalization is required to compare the expression of genes within an individual sample. It can adjust data for two primary technical variables: transcript length and sequencing depth. Longer genes often have more mapped reads than shorter genes at the same expression level. Therefore, their expression level can only be accurately compared within a sample after normalization. Furthermore, the number of sequencing reads per sample may vary. This can also be corrected by within sample normalization. Within sample normalization is not sufficient to compare gene expression between samples. For this, between sample RNA-seq normalization methods are required. Within sample normalization most common techniques are:\n",
    "    - **CPM**: Counts per million (CPM) mapped reads are the number of raw reads mapped to a transcript, scaled by the number of sequencing reads in your sample, multiplied by a million. It normalizes RNA-seq data for sequencing depth but not gene length. Therefore, although it is a within sample normalization approach, CPM normalization is unsuitable for within sample comparisons of gene expression. Between sample comparisons can be made when CPM is used alongside ‘within a dataset’ normalization methods.\n",
    "    - **FPKM/RPKM**: FPKM (fragments per kilobase of transcript per million fragments mapped) for paired-end data and RPKM (reads per kilobase of transcript per million reads mapped) for single-end data correct for variations in library size and gene length. One issue with FPKM/RPKM units is that the expression of a gene in one sample will appear different from its expression in another sample, even when its true expression level is the same. This is because it depends on the relative abundance of a transcript among a population of sequenced transcripts. FPKM/RPKM units best compare gene expression within a single sample\n",
    "    - **TPM**: Transcripts per million (TPM) represents the relative number of transcripts you would detect for a gene if you had sequenced one million full-length transcripts. It is calculated by dividing the number of reads mapped to a transcript by the transcript length. This value is then divided by the sum of mapped reads to all transcripts after normalization for transcript length. It is then multiplied by one million to allow easier further analyses. It normalizes RNA-seq data for sequencing depth and transcript length. TPM and FPKM/RPKM are closely related, however, in contrast to FPKM/RPKM, there is limited variation in values between samples as the sum of all TPMs in each sample is the same. TPM can be used for within sample comparisons but requires ‘within a dataset’ normalization for between sample comparisons.\n",
    "\n",
    " 2. **Within a dataset (between samples)**: Samples within a dataset can be simultaneously normalized as a complete set to adjust for different technical variations such as sequencing depth. RNA-seq is a relative, not an absolute, measure of transcript abundance. This means that the transcript population as a whole affects relative levels of transcripts. This creates biases for gene expression analyses, and these are minimized by between sample RNA-seq normalization methods. Within a dataset normalization mosto common techniques are:\n",
    "    - **Quantile**: The quantile method aims to make the distribution of gene expression levels the same for each sample in a dataset. It assumes that the global differences in distributions between samples are all due to technical variation. Any remaining differences are likely actual biological effects. For each sample, genes are ranked based on their expression level. An average value is calculated across all samples for genes of the same rank. This average value then replaces the original value of all genes in that rank. These genes are then placed in their original order.\n",
    "    - **TMM**: TMM (trimmed mean of M-values) also assumes that most genes are not differentially expressed between samples. If many genes are uniquely or highly expressed in one experimental condition, it will affect the accurate quantification of the remaining genes. To adjust for this possibility, TMM calculates scaling factors to adjust library sizes for the normalization of samples within a dataset. To do this, one sample is chosen as a reference sample. The fold changes and absolute expression levels of other samples within the dataset are then calculated relative to the reference sample. Next, the genes in the data set are ‘trimmed’ to remove differentially expressed genes using these two values. The trimmed mean of the fold changes is then found for each sample. Finally, read counts are scaled by this trimmed mean and the total count of their sample.\n",
    "\n",
    " 3. **Across datasets**: Researchers often integrate RNA-seq data from multiple independent studies. These datasets are usually sequenced at different times, with varying methods across multiple facilities, and contain other experimental variables. This results in a batch effect. The batch effect is often responsible for the greatest source of differential expression when data is combined. It can mask any true biological differences and lead to incorrect conclusions. RNA-seq normalization across datasets can correct for known variables across batches, such as the sequencing center and date of sequencing, as well as unknown variables.\n",
    "\n",
    "source : https://bigomics.ch/blog/why-how-normalize-rna-seq-data/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536b5bb5",
   "metadata": {},
   "source": [
    "**PERSONAL CONSIDERATIONS**\n",
    "\n",
    "Since we want to have a model capable of inferring the survivability of a patient given a feature vector, we thought that the model would need, to behave at best, data normalized across the dataset allowing intra-dataset comparisons in order to learn when a feature vector corresponds to higher or lower survivability. Given this preposition we first normalize miRNA read values with the TMM method to allow genetic intra-dataset comparisons, and then normalize all values of the feature vectors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea29f20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install rnanorm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d0b19a22",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.257776Z",
     "start_time": "2025-10-23T16:22:58.159659Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f47107ac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.288372Z",
     "start_time": "2025-10-23T16:22:59.262349Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "d:\\Universita\\2 anno magistrale\\Progetto BioInf\\miRNA_to_age\n"
     ]
    }
   ],
   "source": [
    "base = os.path.basename(os.getcwd())\n",
    "list = os.getcwd().split(os.sep) \n",
    "list.pop(list.index(base))\n",
    "ROOT = '\\\\'.join(list)\n",
    "print(ROOT)\n",
    "DATA_PATH = os.path.join(ROOT, 'datasets\\\\preprocessed')\n",
    "miRNA_file = 'clinical_miRNA.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "33b0f96b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.319251Z",
     "start_time": "2025-10-23T16:22:59.292386Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d:\\\\Universita\\\\2 anno magistrale\\\\Progetto BioInf\\\\miRNA_to_age\\\\datasets\\\\preprocessed\\\\clinical_miRNA.csv'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miRNA_dataset = os.path.join(DATA_PATH, miRNA_file)\n",
    "miRNA_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a9377bf2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.769799Z",
     "start_time": "2025-10-23T16:22:59.326271Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_followup</th>\n",
       "      <th>case_id</th>\n",
       "      <th>read_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>71.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1918.0</td>\n",
       "      <td>378778d2-b331-4867-a93b-c64028c8b4c7</td>\n",
       "      <td>[13756, 13807, 13949, 55698, 5797, 518, 3747, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>53.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1309.0</td>\n",
       "      <td>b343bfe0-7c23-4c6a-8c84-9ee39db2ecda</td>\n",
       "      <td>[37711, 37303, 37662, 44231, 14405, 1889, 3169...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>55.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>abdc76db-f85e-4337-a57e-6d098789da03</td>\n",
       "      <td>[10731, 10926, 10792, 14125, 2622, 462, 698, 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>64.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>212.0</td>\n",
       "      <td>fbee40f1-d6d8-4156-8d42-36e09bb9f095</td>\n",
       "      <td>[70280, 70637, 70972, 67833, 5167, 1629, 18954...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>46.0</td>\n",
       "      <td>2763.0</td>\n",
       "      <td>2763.0</td>\n",
       "      <td>50619f8c-10aa-464a-a227-90a7aa6ffd43</td>\n",
       "      <td>[12807, 12718, 13064, 26120, 4490, 1775, 11051...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   age_at_initial_pathologic_diagnosis  days_to_death  days_to_last_followup  \\\n",
       "0                                 71.0           -1.0                 1918.0   \n",
       "1                                 53.0           -1.0                 1309.0   \n",
       "2                                 55.0           -1.0                    0.0   \n",
       "3                                 64.0           -1.0                  212.0   \n",
       "4                                 46.0         2763.0                 2763.0   \n",
       "\n",
       "                                case_id  \\\n",
       "0  378778d2-b331-4867-a93b-c64028c8b4c7   \n",
       "1  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda   \n",
       "2  abdc76db-f85e-4337-a57e-6d098789da03   \n",
       "3  fbee40f1-d6d8-4156-8d42-36e09bb9f095   \n",
       "4  50619f8c-10aa-464a-a227-90a7aa6ffd43   \n",
       "\n",
       "                                          read_count  \n",
       "0  [13756, 13807, 13949, 55698, 5797, 518, 3747, ...  \n",
       "1  [37711, 37303, 37662, 44231, 14405, 1889, 3169...  \n",
       "2  [10731, 10926, 10792, 14125, 2622, 462, 698, 3...  \n",
       "3  [70280, 70637, 70972, 67833, 5167, 1629, 18954...  \n",
       "4  [12807, 12718, 13064, 26120, 4490, 1775, 11051...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(miRNA_dataset)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772fa321",
   "metadata": {},
   "source": [
    "## Data preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "338bbaa3",
   "metadata": {},
   "source": [
    "creation of file with all miRNA base names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bd926b90",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.800534Z",
     "start_time": "2025-10-23T16:22:59.773802Z"
    }
   },
   "outputs": [],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'genes_names.txt')):\n",
    "    with open(os.path.join(DATA_PATH, 'genes_names.txt'), 'r') as f: \n",
    "        lines = f.readlines()\n",
    "        genes_names = [line.strip() for line in lines]\n",
    "else:\n",
    "    genes_names = []\n",
    "    #path = 'D:\\\\Universita\\\\2 anno magistrale\\\\Progetto BioInf\\\\miRNA_to_age\\\\datasets\\\\miRNA_seq\\\\0b64742b-b99d-4fc1-bfbf-bb7074f21a67\\\\9293a93f-b3d1-41ab-8c18-3deed4cc776d.mirbase21.mirnas.quantification.txt'\n",
    "    path = '../datasets/miRNA_seq/0b64742b-b99d-4fc1-bfbf-bb7074f21a67/9293a93f-b3d1-41ab-8c18-3deed4cc776d.mirbase21.mirnas.quantification.txt'\n",
    "    with open(path, 'r') as f: \n",
    "        lines = f.readlines()[1:]\n",
    "        for l in lines:\n",
    "            genes_names.append(l.split('\\t')[0])\n",
    "\n",
    "    with open(os.path.join(DATA_PATH, 'genes_names.txt'), 'w') as f:\n",
    "        for name in genes_names:\n",
    "            f.write(name + '\\n') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec543277",
   "metadata": {},
   "source": [
    "extraction of miRNA_reads and adjustment into dataframe with each column being a gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2fa38438",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:40:31.312716Z",
     "start_time": "2025-10-23T18:40:31.262718Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c  \\\n",
      "Sample_0       13756.0       13807.0       13949.0     55698.0      5797.0   \n",
      "Sample_1       37711.0       37303.0       37662.0     44231.0     14405.0   \n",
      "Sample_2       10731.0       10926.0       10792.0     14125.0      2622.0   \n",
      "Sample_3       70280.0       70637.0       70972.0     67833.0      5167.0   \n",
      "Sample_4       12807.0       12718.0       13064.0     26120.0      4490.0   \n",
      "\n",
      "          hsa-let-7d  hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  hsa-let-7g  ...  \\\n",
      "Sample_0       518.0      3747.0        4192.0        4212.0       318.0  ...   \n",
      "Sample_1      1889.0      3169.0       27465.0       27650.0      1935.0  ...   \n",
      "Sample_2       462.0       698.0        3780.0        3783.0       438.0  ...   \n",
      "Sample_3      1629.0     18954.0       31984.0       32414.0      2036.0  ...   \n",
      "Sample_4      1775.0     11051.0        5293.0        5345.0      2075.0  ...   \n",
      "\n",
      "          hsa-mir-941-5  hsa-mir-942  hsa-mir-943  hsa-mir-944  hsa-mir-95  \\\n",
      "Sample_0            0.0          4.0          0.0          0.0         2.0   \n",
      "Sample_1            0.0          9.0          0.0          4.0         3.0   \n",
      "Sample_2            0.0          9.0          0.0          0.0         2.0   \n",
      "Sample_3            0.0         19.0          0.0          3.0        11.0   \n",
      "Sample_4            0.0         11.0          0.0          0.0        12.0   \n",
      "\n",
      "          hsa-mir-9500  hsa-mir-96  hsa-mir-98  hsa-mir-99a  hsa-mir-99b  \n",
      "Sample_0           0.0         3.0       135.0        839.0     101810.0  \n",
      "Sample_1           0.0        49.0       249.0       2499.0      53660.0  \n",
      "Sample_2           0.0        16.0        33.0       1536.0      39243.0  \n",
      "Sample_3           0.0       183.0       593.0       1567.0     259364.0  \n",
      "Sample_4           0.0       178.0       234.0       1278.0     437321.0  \n",
      "\n",
      "[5 rows x 1881 columns]\n"
     ]
    }
   ],
   "source": [
    "def parse_array(x):\n",
    "    if isinstance(x, str):\n",
    "        x = x.strip(\"[]\")\n",
    "        return np.array([float(i) for i in x.split(\",\")])\n",
    "    \n",
    "    return np.array(x)\n",
    "\n",
    "if os.path.exists(os.path.join(DATA_PATH, 'genes_reads.csv')):\n",
    "    with open(os.path.join(DATA_PATH, 'genes_reads.csv'), 'r') as f:\n",
    "        genes_reads = pd.read_csv(os.path.join(DATA_PATH, 'genes_reads.csv'), index_col=0)\n",
    "\n",
    "else:\n",
    "    n_cols = len(data[\"read_count\"][0].strip(\"[]\").split(\",\"))\n",
    "    reads = data[\"read_count\"].apply(parse_array)\n",
    "\n",
    "    genes_reads = pd.DataFrame(\n",
    "        np.stack(reads.values),  # Converte la Series di vettori in un array 2D\n",
    "        index=[f'Sample_{i}' for i in data.index]\n",
    "    )\n",
    "\n",
    "    genes_reads.columns = genes_names\n",
    "\n",
    "    genes_reads.to_csv(os.path.join(DATA_PATH, 'genes_reads.csv'))\n",
    "    print(genes_reads.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f70f7e",
   "metadata": {},
   "source": [
    "### Extraction from web and manual dictionary of the corresponendce between miRNA base names and gene_id and gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cf423c59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              gene_length\n",
      "gene_id                  \n",
      "hsa-let-7a-1           80\n",
      "hsa-let-7a-2           72\n",
      "hsa-let-7a-3           74\n",
      "hsa-let-7b             83\n",
      "hsa-let-7c             84\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'gene_lengths.csv')):\n",
    "    gene_lengths_df = pd.read_csv(os.path.join(DATA_PATH, 'gene_lengths.csv'), sep=\"\\t\", index_col=0)\n",
    "    print(gene_lengths_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a1ce1344",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:21.270091Z",
     "start_time": "2025-10-23T16:23:20.985272Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Gene stable ID  Gene name      miRBase ID\n",
      "0  ENSG00000283344  MIR1244-4  hsa-mir-1244-1\n",
      "1  ENSG00000283344  MIR1244-4  hsa-mir-1244-2\n",
      "2  ENSG00000283344  MIR1244-4  hsa-mir-1244-3\n",
      "3  ENSG00000283344  MIR1244-4  hsa-mir-1244-4\n",
      "4  ENSG00000292346    MIR6089  hsa-mir-6089-1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2170, 3)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pybiomart import Server\n",
    "\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "mart = server['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = mart['hsapiens_gene_ensembl']\n",
    "\n",
    "# Query con filtro sui geni miRNA\n",
    "miRNA_mapping = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name', 'mirbase_id'],\n",
    "    filters={'biotype': 'miRNA'}  # filtra solo i miRNA\n",
    ")\n",
    "\n",
    "print(miRNA_mapping.head())\n",
    "miRNA_mapping.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b1d8d3161cf5e0a9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:23.906180Z",
     "start_time": "2025-10-23T16:23:23.677326Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene stable ID Gene name    miRBase ID\n",
      "2   ENSG00000252695   MIR2276  hsa-mir-2276\n",
      "5   ENSG00000263399   MIR3170  hsa-mir-3170\n",
      "7   ENSG00000207719    MIR623   hsa-mir-623\n",
      "9   ENSG00000263615   MIR4306  hsa-mir-4306\n",
      "12  ENSG00000265164   MIR2681  hsa-mir-2681\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1601, 3)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server = Server(host='http://grch37.ensembl.org')\n",
    "mart = server['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = mart['hsapiens_gene_ensembl']\n",
    "\n",
    "mapping_grch37 = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name', 'mirbase_id'],\n",
    "    filters={'biotype': 'miRNA'}\n",
    ")\n",
    "miRNA_mapping_grch37 = mapping_grch37[mapping_grch37['miRBase ID'].notnull()].drop_duplicates()\n",
    "\n",
    "print(miRNA_mapping_grch37.head())\n",
    "miRNA_mapping_grch37.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b9e2e188",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:26.053746Z",
     "start_time": "2025-10-23T16:23:25.998692Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature  start    end score strand frame  \\\n",
       "0    chr1  ENSEMBL        gene  17369  17436     .      -     .   \n",
       "1    chr1  ENSEMBL  transcript  17369  17436     .      -     .   \n",
       "2    chr1  ENSEMBL        exon  17369  17436     .      -     .   \n",
       "3    chr1  ENSEMBL        gene  30366  30503     .      +     .   \n",
       "4    chr1  ENSEMBL  transcript  30366  30503     .      +     .   \n",
       "\n",
       "             gene_id gene_type  ... exon_number            exon_id  \\\n",
       "0  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n",
       "1  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n",
       "2  ENSG00000278267.1     miRNA  ...         1.0  ENSE00003746039.1   \n",
       "3  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n",
       "4  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n",
       "\n",
       "  transcript_support_level havana_transcript     hgnc_id havana_gene  ont  \\\n",
       "0                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "1                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "2                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "3                      NaN               NaN  HGNC:35294         NaN  NaN   \n",
       "4                      NaN               NaN  HGNC:35294         NaN  NaN   \n",
       "\n",
       "  protein_id  ccdsid  artif_dupl  \n",
       "0        NaN     NaN         NaN  \n",
       "1        NaN     NaN         NaN  \n",
       "2        NaN     NaN         NaN  \n",
       "3        NaN     NaN         NaN  \n",
       "4        NaN     NaN         NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_miRNA = pd.read_csv(os.path.join(DATA_PATH, 'miRNA_gtf.csv'))\n",
    "gencode_miRNA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c629d1c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:27:12.906840Z",
     "start_time": "2025-10-23T18:23:45.451456Z"
    }
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[15]\u001b[39m\u001b[32m, line 31\u001b[39m\n\u001b[32m     29\u001b[39m result = pd.DataFrame()\n\u001b[32m     30\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m gene_id_value:\n\u001b[32m---> \u001b[39m\u001b[32m31\u001b[39m     result = gencode_miRNA[\u001b[43mgencode_miRNA\u001b[49m\u001b[43m.\u001b[49m\u001b[43mastype\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m.\u001b[49m\u001b[43mstr\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcontains\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene_id_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcase\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m.any(axis=\u001b[32m1\u001b[39m)]\n\u001b[32m     33\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m result.empty:\n\u001b[32m     34\u001b[39m     \u001b[38;5;66;03m# Result is empty, trying with Gene stable ID\u001b[39;00m\n\u001b[32m     35\u001b[39m     gene_stable_id = \u001b[38;5;28mmap\u001b[39m[\u001b[33m'\u001b[39m\u001b[33mGene stable ID\u001b[39m\u001b[33m'\u001b[39m].iloc[\u001b[32m0\u001b[39m]\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\frame.py:10381\u001b[39m, in \u001b[36mDataFrame.apply\u001b[39m\u001b[34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001b[39m\n\u001b[32m  10367\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpandas\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mcore\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mapply\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m frame_apply\n\u001b[32m  10369\u001b[39m op = frame_apply(\n\u001b[32m  10370\u001b[39m     \u001b[38;5;28mself\u001b[39m,\n\u001b[32m  10371\u001b[39m     func=func,\n\u001b[32m   (...)\u001b[39m\u001b[32m  10379\u001b[39m     kwargs=kwargs,\n\u001b[32m  10380\u001b[39m )\n\u001b[32m> \u001b[39m\u001b[32m10381\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mop\u001b[49m\u001b[43m.\u001b[49m\u001b[43mapply\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m.__finalize__(\u001b[38;5;28mself\u001b[39m, method=\u001b[33m\"\u001b[39m\u001b[33mapply\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\apply.py:916\u001b[39m, in \u001b[36mFrameApply.apply\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    913\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.raw:\n\u001b[32m    914\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m.apply_raw(engine=\u001b[38;5;28mself\u001b[39m.engine, engine_kwargs=\u001b[38;5;28mself\u001b[39m.engine_kwargs)\n\u001b[32m--> \u001b[39m\u001b[32m916\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mapply_standard\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\apply.py:1063\u001b[39m, in \u001b[36mFrameApply.apply_standard\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1061\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mapply_standard\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[32m   1062\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.engine == \u001b[33m\"\u001b[39m\u001b[33mpython\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m-> \u001b[39m\u001b[32m1063\u001b[39m         results, res_index = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mapply_series_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1064\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1065\u001b[39m         results, res_index = \u001b[38;5;28mself\u001b[39m.apply_series_numba()\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\apply.py:1081\u001b[39m, in \u001b[36mFrameApply.apply_series_generator\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1078\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m option_context(\u001b[33m\"\u001b[39m\u001b[33mmode.chained_assignment\u001b[39m\u001b[33m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[32m   1079\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m i, v \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(series_gen):\n\u001b[32m   1080\u001b[39m         \u001b[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1081\u001b[39m         results[i] = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1082\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(results[i], ABCSeries):\n\u001b[32m   1083\u001b[39m             \u001b[38;5;66;03m# If we have a view on v, we need to make a copy because\u001b[39;00m\n\u001b[32m   1084\u001b[39m             \u001b[38;5;66;03m#  series_generator will swap out the underlying data\u001b[39;00m\n\u001b[32m   1085\u001b[39m             results[i] = results[i].copy(deep=\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[15]\u001b[39m\u001b[32m, line 31\u001b[39m, in \u001b[36m<lambda>\u001b[39m\u001b[34m(x)\u001b[39m\n\u001b[32m     29\u001b[39m result = pd.DataFrame()\n\u001b[32m     30\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m gene_id_value:\n\u001b[32m---> \u001b[39m\u001b[32m31\u001b[39m     result = gencode_miRNA[gencode_miRNA.astype(\u001b[38;5;28mstr\u001b[39m).apply(\u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[43mx\u001b[49m\u001b[43m.\u001b[49m\u001b[43mstr\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcontains\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene_id_value\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcase\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m).any(axis=\u001b[32m1\u001b[39m)]\n\u001b[32m     33\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m result.empty:\n\u001b[32m     34\u001b[39m     \u001b[38;5;66;03m# Result is empty, trying with Gene stable ID\u001b[39;00m\n\u001b[32m     35\u001b[39m     gene_stable_id = \u001b[38;5;28mmap\u001b[39m[\u001b[33m'\u001b[39m\u001b[33mGene stable ID\u001b[39m\u001b[33m'\u001b[39m].iloc[\u001b[32m0\u001b[39m]\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\strings\\accessor.py:140\u001b[39m, in \u001b[36mforbid_nonstring_types.<locals>._forbid_nonstring_types.<locals>.wrapper\u001b[39m\u001b[34m(self, *args, **kwargs)\u001b[39m\n\u001b[32m    135\u001b[39m     msg = (\n\u001b[32m    136\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mCannot use .str.\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m with values of \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    137\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33minferred dtype \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m._inferred_dtype\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    138\u001b[39m     )\n\u001b[32m    139\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg)\n\u001b[32m--> \u001b[39m\u001b[32m140\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\strings\\accessor.py:1346\u001b[39m, in \u001b[36mStringMethods.contains\u001b[39m\u001b[34m(self, pat, case, flags, na, regex)\u001b[39m\n\u001b[32m   1338\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m regex \u001b[38;5;129;01mand\u001b[39;00m re.compile(pat).groups:\n\u001b[32m   1339\u001b[39m     warnings.warn(\n\u001b[32m   1340\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mThis pattern is interpreted as a regular expression, and has \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1341\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mmatch groups. To actually get the groups, use str.extract.\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m   1342\u001b[39m         \u001b[38;5;167;01mUserWarning\u001b[39;00m,\n\u001b[32m   1343\u001b[39m         stacklevel=find_stack_level(),\n\u001b[32m   1344\u001b[39m     )\n\u001b[32m-> \u001b[39m\u001b[32m1346\u001b[39m result = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_data\u001b[49m\u001b[43m.\u001b[49m\u001b[43marray\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_str_contains\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpat\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcase\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mregex\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1347\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._wrap_result(result, fill_value=na, returns_string=\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\strings\\object_array.py:162\u001b[39m, in \u001b[36mObjectStringArrayMixin._str_contains\u001b[39m\u001b[34m(self, pat, case, flags, na, regex)\u001b[39m\n\u001b[32m    154\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m na \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m lib.no_default \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m isna(na) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(na, \u001b[38;5;28mbool\u001b[39m):\n\u001b[32m    155\u001b[39m     \u001b[38;5;66;03m# GH#59561\u001b[39;00m\n\u001b[32m    156\u001b[39m     warnings.warn(\n\u001b[32m    157\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mAllowing a non-bool \u001b[39m\u001b[33m'\u001b[39m\u001b[33mna\u001b[39m\u001b[33m'\u001b[39m\u001b[33m in obj.str.contains is deprecated \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    158\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mand will raise in a future version.\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m    159\u001b[39m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[32m    160\u001b[39m         stacklevel=find_stack_level(),\n\u001b[32m    161\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m162\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_str_map\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mbool\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\pandas\\core\\strings\\object_array.py:107\u001b[39m, in \u001b[36mObjectStringArrayMixin._str_map\u001b[39m\u001b[34m(self, f, na_value, dtype, convert)\u001b[39m\n\u001b[32m    105\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[32m    106\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m na_value \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m np.nan:\n\u001b[32m--> \u001b[39m\u001b[32m107\u001b[39m     \u001b[43mnp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mputmask\u001b[49m\u001b[43m(\u001b[49m\u001b[43mresult\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_value\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    108\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m convert \u001b[38;5;129;01mand\u001b[39;00m result.dtype == \u001b[38;5;28mobject\u001b[39m:\n\u001b[32m    109\u001b[39m         result = lib.maybe_convert_objects(result)\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "manual_mapping = {\n",
    "    \"hsa-mir-3607\": \"MIR3607\",\n",
    "    \"hsa-mir-3653\": \"MIR3653\",\n",
    "    \"hsa-mir-3687-1\": \"MIR3687-1\",\n",
    "    \"hsa-mir-3687-2\": \"MIR3687-2\",\n",
    "    \"hsa-mir-6087\": \"MIR6087\",\n",
    "    \"hsa-mir-6723\": \"MIR6723\",\n",
    "    \"hsa-mir-6827\": \"MIR6827\",\n",
    "    \"hsa-mir-7641-1\": \"MIR7641-1\",\n",
    "    \"hsa-mir-7641-2\": \"MIR7641-2\",\n",
    "    \"hsa-mir-3656\": \"MIR3656\",\n",
    "    \"hsa-mir-4788\": \"MIR4788\"\n",
    "}\n",
    "\n",
    "gene_lengths = {}\n",
    "for name in genes_names:\n",
    "    gene_id_value = None\n",
    "    if name in manual_mapping:\n",
    "        gene_id_value = manual_mapping[name]\n",
    "    else:\n",
    "        map = miRNA_mapping[miRNA_mapping.astype(str).apply(lambda x: x.str.contains(name, case=False, na=False)).any(axis=1)]\n",
    "        if map.empty:\n",
    "            map = miRNA_mapping_grch37[miRNA_mapping_grch37.astype(str).apply(lambda x: x.str.contains(name, case=False, na=False)).any(axis=1)]\n",
    "        if map.empty:\n",
    "            print(f\"Gene not found for miRBase ID: {name}\")\n",
    "            continue            \n",
    "        gene_id_value = map['Gene name'].iloc[0] if not map['Gene name'].empty else None\n",
    "    \n",
    "    result = pd.DataFrame()\n",
    "    if gene_id_value:\n",
    "        result = gencode_miRNA[gencode_miRNA.astype(str).apply(lambda x: x.str.contains(gene_id_value, case=False, na=False)).any(axis=1)]\n",
    "    \n",
    "    if result.empty:\n",
    "        # Result is empty, trying with Gene stable ID\n",
    "        gene_stable_id = map['Gene stable ID'].iloc[0]\n",
    "        result = gencode_miRNA[gencode_miRNA.astype(str).apply(lambda x: x.str.contains(gene_stable_id, case=False, na=False)).any(axis=1)]\n",
    "        \n",
    "        if result.empty:\n",
    "            print(f\"Gene not found for miRBase ID: {name} using both Gene name and Gene stable ID\")\n",
    "            continue\n",
    "      \n",
    "    gene_lengths[name] = int(result['end'].values[0]) - int(result['start'].values[0]) + 1\n",
    "\n",
    "gene_lengths_df = pd.DataFrame([\n",
    "    {'gene_id': name, 'gene_length': length}\n",
    "    for name, length in gene_lengths.items()\n",
    "])\n",
    "\n",
    "print(gene_lengths_df.head())\n",
    "print(gene_lengths_df.shape)\n",
    "print(gene_lengths_df.isna().sum())\n",
    "gene_lengths_df.to_csv(\"gene_lengths.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acd9e60befa03895",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "The manual mapping was made consulting the miRBase database at https://www.mirbase.org/ ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ec42e51",
   "metadata": {},
   "source": [
    "### Quantile normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0b71cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c\n",
      "Sample_0       13756.0       13807.0       13949.0     55698.0      5797.0\n",
      "Sample_1       37711.0       37303.0       37662.0     44231.0     14405.0\n",
      "Sample_2       10731.0       10926.0       10792.0     14125.0      2622.0\n",
      "Sample_3       70280.0       70637.0       70972.0     67833.0      5167.0\n",
      "Sample_4       12807.0       12718.0       13064.0     26120.0      4490.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Matteo Bulgarelli\\AppData\\Local\\Temp\\ipykernel_16608\\2586111674.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  reads_logged = reads.applymap(lambda x: np.log2(x + 1))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c\n",
      "Sample_0     13.747878     13.753217     13.767978   15.765364   12.501340\n",
      "Sample_1     15.202736     15.187043     15.200860   15.432803   13.814382\n",
      "Sample_2     13.389631     13.415610     13.397808   13.786065   11.357002\n",
      "Sample_3     16.100847     16.108157     16.114983   16.049721   12.335390\n",
      "Sample_4     13.644758     13.634698     13.673420   14.672923   12.132821\n",
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c\n",
      "Sample_0      1.152513      1.153766      1.150609    1.331246    1.488567\n",
      "Sample_1      1.660901      1.659254      1.656288    1.242474    2.112232\n",
      "Sample_2      1.056369      1.064106      1.055602    0.837121    1.153766\n",
      "Sample_3      2.123158      2.117345      2.123158    1.451740    1.433827\n",
      "Sample_4      1.134205      1.134205      1.134205    1.030723    1.381485\n"
     ]
    }
   ],
   "source": [
    "import qnorm\n",
    "\n",
    "reads = genes_reads\n",
    "print(reads.iloc[:5, :5])\n",
    "\n",
    "# Log2 transform necessary for 0 values before quantile normalization\n",
    "reads_logged = reads.applymap(lambda x: np.log2(x + 1))\n",
    "print(reads_logged.iloc[:5, :5])\n",
    "\n",
    "# Quantile normalization\n",
    "quant_norm = qnorm.quantile_normalize(reads_logged)\n",
    "print(quant_norm.iloc[:5, :5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b955782e",
   "metadata": {},
   "source": [
    "### TMM normalization (NOT USED because not reccomended for miRNA-seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "e5242d6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c\n",
      "Sample_0       13756.0       13807.0       13949.0     55698.0      5797.0\n",
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3    hsa-let-7b   hsa-let-7c\n",
      "Sample_0  14437.364984  14490.891127    14639.9247  58456.844643  6084.138181\n"
     ]
    }
   ],
   "source": [
    "import conorm\n",
    "\n",
    "reads = genes_reads.T\n",
    "\n",
    "conorm_reads = conorm.tmm(reads)\n",
    "# print(conorm_reads.T.head())\n",
    "\n",
    "print(genes_reads.iloc[:1, :5])\n",
    "print(conorm_reads.T.iloc[:1, :5])\n",
    "# print(conorm.tmm_norm_factors(conorm_reads))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91ca679",
   "metadata": {},
   "source": [
    "### Creation of normalized file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "d180a663",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   days_to_death pathologic_stage  age_at_initial_pathologic_diagnosis  \\\n",
      "0           -1.0        Stage IIA                                   71   \n",
      "1           -1.0          Stage I                                   53   \n",
      "2           -1.0        Stage IIB                                   55   \n",
      "3           -1.0        Stage IIA                                   64   \n",
      "4         2763.0          Stage I                                   46   \n",
      "\n",
      "   days_to_last_followup  Death                               case_id  \\\n",
      "0                 1918.0      0  378778d2-b331-4867-a93b-c64028c8b4c7   \n",
      "1                 1309.0      0  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda   \n",
      "2                    0.0      0  abdc76db-f85e-4337-a57e-6d098789da03   \n",
      "3                  212.0      0  fbee40f1-d6d8-4156-8d42-36e09bb9f095   \n",
      "4                 2763.0      1  50619f8c-10aa-464a-a227-90a7aa6ffd43   \n",
      "\n",
      "                                          read_count  \n",
      "0  [13756, 13807, 13949, 55698, 5797, 518, 3747, ...  \n",
      "1  [37711, 37303, 37662, 44231, 14405, 1889, 3169...  \n",
      "2  [10731, 10926, 10792, 14125, 2622, 462, 698, 3...  \n",
      "3  [70280, 70637, 70972, 67833, 5167, 1629, 18954...  \n",
      "4  [12807, 12718, 13064, 26120, 4490, 1775, 11051...  \n",
      "   days_to_death pathologic_stage  age_at_initial_pathologic_diagnosis  \\\n",
      "0           -1.0        Stage IIA                                   71   \n",
      "1           -1.0          Stage I                                   53   \n",
      "2           -1.0        Stage IIB                                   55   \n",
      "3           -1.0        Stage IIA                                   64   \n",
      "4         2763.0          Stage I                                   46   \n",
      "\n",
      "   days_to_last_followup  Death                               case_id  \\\n",
      "0                 1918.0      0  378778d2-b331-4867-a93b-c64028c8b4c7   \n",
      "1                 1309.0      0  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda   \n",
      "2                    0.0      0  abdc76db-f85e-4337-a57e-6d098789da03   \n",
      "3                  212.0      0  fbee40f1-d6d8-4156-8d42-36e09bb9f095   \n",
      "4                 2763.0      1  50619f8c-10aa-464a-a227-90a7aa6ffd43   \n",
      "\n",
      "                                          read_count  \n",
      "0  [1.1525132336053339, 1.1537659472457886, 1.150...  \n",
      "1  [1.6609008116263617, 1.659253920825836, 1.6562...  \n",
      "2  [1.056368801368503, 1.0641055787474447, 1.0556...  \n",
      "3  [2.1231578536606395, 2.1173454338208257, 2.123...  \n",
      "4  [1.1342047472759338, 1.1342047472759338, 1.134...  \n"
     ]
    }
   ],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'clinical_miRNA_normalized.csv')):\n",
    "    normalized = pd.read_csv(os.path.join(DATA_PATH, 'clinical_miRNA_normalized.csv'))\n",
    "    print(normalized.head())\n",
    "\n",
    "else:\n",
    "    final_merged_miRNA = pd.DataFrame(\n",
    "        columns=['read_count'],\n",
    "        data=quant_norm.apply(lambda row: row.values, axis=1)\n",
    "    )\n",
    "    # print(final_merged_miRNA.head())\n",
    "    # print(final_merged_miRNA.shape)\n",
    "\n",
    "    normalized = pd.read_csv(miRNA_dataset)\n",
    "    print(normalized.head())\n",
    "    normalized['read_count'] = final_merged_miRNA['read_count'].values\n",
    "    print(normalized.head())\n",
    "\n",
    "    normalized.to_csv(os.path.join(DATA_PATH, 'clinical_miRNA_normalized.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tirocinio-venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
