{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bda669cc",
   "metadata": {},
   "source": [
    "# DATA NORMALIZATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b436641",
   "metadata": {},
   "source": [
    "Various factors affect transcript quantification in RNA-seq data, such as sequencing depth, transcript length, and sample-to-sample and batch-to-batch variability. Normalization methods exist to minimize these variables and ensure reliable transcriptomic data. Sequencing technologies introduce technical variability. Therefore, raw transcriptomic data must be adjusted to account for these technical factors. \n",
    "\n",
    "It is essential to choose the correct RNA-seq normalization method for the dataset and there are three main RNA-seq normalization stages to consider:\n",
    " 1. **Within sample**: Within sample normalization is required to compare the expression of genes within an individual sample. It can adjust data for two primary technical variables: transcript length and sequencing depth. Longer genes often have more mapped reads than shorter genes at the same expression level. Therefore, their expression level can only be accurately compared within a sample after normalization. Furthermore, the number of sequencing reads per sample may vary. This can also be corrected by within sample normalization. Within sample normalization is not sufficient to compare gene expression between samples. For this, between sample RNA-seq normalization methods are required. Within sample normalization most common techniques are:\n",
    "    - **CPM**: Counts per million (CPM) mapped reads are the number of raw reads mapped to a transcript, scaled by the number of sequencing reads in your sample, multiplied by a million. It normalizes RNA-seq data for sequencing depth but not gene length. Therefore, although it is a within sample normalization approach, CPM normalization is unsuitable for within sample comparisons of gene expression. Between sample comparisons can be made when CPM is used alongside ‘within a dataset’ normalization methods.\n",
    "    - **FPKM/RPKM**: FPKM (fragments per kilobase of transcript per million fragments mapped) for paired-end data and RPKM (reads per kilobase of transcript per million reads mapped) for single-end data correct for variations in library size and gene length. One issue with FPKM/RPKM units is that the expression of a gene in one sample will appear different from its expression in another sample, even when its true expression level is the same. This is because it depends on the relative abundance of a transcript among a population of sequenced transcripts. FPKM/RPKM units best compare gene expression within a single sample\n",
    "    - **TPM**: Transcripts per million (TPM) represents the relative number of transcripts you would detect for a gene if you had sequenced one million full-length transcripts. It is calculated by dividing the number of reads mapped to a transcript by the transcript length. This value is then divided by the sum of mapped reads to all transcripts after normalization for transcript length. It is then multiplied by one million to allow easier further analyses. It normalizes RNA-seq data for sequencing depth and transcript length. TPM and FPKM/RPKM are closely related, however, in contrast to FPKM/RPKM, there is limited variation in values between samples as the sum of all TPMs in each sample is the same. TPM can be used for within sample comparisons but requires ‘within a dataset’ normalization for between sample comparisons.\n",
    "\n",
    " 2. **Within a dataset (between samples)**: Samples within a dataset can be simultaneously normalized as a complete set to adjust for different technical variations such as sequencing depth. RNA-seq is a relative, not an absolute, measure of transcript abundance. This means that the transcript population as a whole affects relative levels of transcripts. This creates biases for gene expression analyses, and these are minimized by between sample RNA-seq normalization methods. Within a dataset normalization mosto common techniques are:\n",
    "    - **Quantile**: The quantile method aims to make the distribution of gene expression levels the same for each sample in a dataset. It assumes that the global differences in distributions between samples are all due to technical variation. Any remaining differences are likely actual biological effects. For each sample, genes are ranked based on their expression level. An average value is calculated across all samples for genes of the same rank. This average value then replaces the original value of all genes in that rank. These genes are then placed in their original order.\n",
    "    - **TMM**: TMM (trimmed mean of M-values) also assumes that most genes are not differentially expressed between samples. If many genes are uniquely or highly expressed in one experimental condition, it will affect the accurate quantification of the remaining genes. To adjust for this possibility, TMM calculates scaling factors to adjust library sizes for the normalization of samples within a dataset. To do this, one sample is chosen as a reference sample. The fold changes and absolute expression levels of other samples within the dataset are then calculated relative to the reference sample. Next, the genes in the data set are ‘trimmed’ to remove differentially expressed genes using these two values. The trimmed mean of the fold changes is then found for each sample. Finally, read counts are scaled by this trimmed mean and the total count of their sample.\n",
    "\n",
    " 3. **Across datasets**: Researchers often integrate RNA-seq data from multiple independent studies. These datasets are usually sequenced at different times, with varying methods across multiple facilities, and contain other experimental variables. This results in a batch effect. The batch effect is often responsible for the greatest source of differential expression when data is combined. It can mask any true biological differences and lead to incorrect conclusions. RNA-seq normalization across datasets can correct for known variables across batches, such as the sequencing center and date of sequencing, as well as unknown variables.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536b5bb5",
   "metadata": {},
   "source": [
    "**PERSONAL CONSIDERATIONS**\n",
    "\n",
    "We didn't think the Across Dataset normalization was needed in our case since all samples came from same laboratory and all samples were equal in terms of desease and in general terms. WE thought pf applying only the first 2 types of normalization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea29f20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install rnanorm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d0b19a22",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.257776Z",
     "start_time": "2025-10-23T16:22:58.159659Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f47107ac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.288372Z",
     "start_time": "2025-10-23T16:22:59.262349Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "d:\\Universita\\2 anno magistrale\\Progetto BioInf\\miRNA_to_age\n"
     ]
    }
   ],
   "source": [
    "base = os.path.basename(os.getcwd())\n",
    "list = os.getcwd().split(os.sep) \n",
    "list.pop(list.index(base))\n",
    "ROOT = '\\\\'.join(list)\n",
    "print(ROOT)\n",
    "DATA_PATH = os.path.join(ROOT, 'datasets\\\\preprocessed')\n",
    "miRNA_file = 'final_merged_clinical_miRNA.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "33b0f96b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.319251Z",
     "start_time": "2025-10-23T16:22:59.292386Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'d:\\\\Universita\\\\2 anno magistrale\\\\Progetto BioInf\\\\miRNA_to_age\\\\datasets\\\\preprocessed\\\\final_merged.csv'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "miRNA_dataset = os.path.join(DATA_PATH, miRNA_file)\n",
    "miRNA_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a9377bf2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.769799Z",
     "start_time": "2025-10-23T16:22:59.326271Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>reads_per_million_miRNA_mapped</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>378778d2-b331-4867-a93b-c64028c8b4c7</td>\n",
       "      <td>71.0</td>\n",
       "      <td>[12120.990742, 12041.557881, 12141.51573, 1768...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b343bfe0-7c23-4c6a-8c84-9ee39db2ecda</td>\n",
       "      <td>53.0</td>\n",
       "      <td>[20904.21904, 21106.26856, 20907.030164, 45794...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b343bfe0-7c23-4c6a-8c84-9ee39db2ecda</td>\n",
       "      <td>53.0</td>\n",
       "      <td>[4742.042712, 4748.951756, 4919.239371, 4809.9...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3e775c99-ceda-4246-8d6f-0f58ca5097c8</td>\n",
       "      <td>59.0</td>\n",
       "      <td>[8992.901265, 8985.700823, 9044.654437, 29490....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>abdc76db-f85e-4337-a57e-6d098789da03</td>\n",
       "      <td>55.0</td>\n",
       "      <td>[6540.401419, 6668.261153, 6475.719437, 19448....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id  age_at_initial_pathologic_diagnosis  \\\n",
       "0  378778d2-b331-4867-a93b-c64028c8b4c7                                 71.0   \n",
       "1  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda                                 53.0   \n",
       "2  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda                                 53.0   \n",
       "3  3e775c99-ceda-4246-8d6f-0f58ca5097c8                                 59.0   \n",
       "4  abdc76db-f85e-4337-a57e-6d098789da03                                 55.0   \n",
       "\n",
       "                      reads_per_million_miRNA_mapped  \n",
       "0  [12120.990742, 12041.557881, 12141.51573, 1768...  \n",
       "1  [20904.21904, 21106.26856, 20907.030164, 45794...  \n",
       "2  [4742.042712, 4748.951756, 4919.239371, 4809.9...  \n",
       "3  [8992.901265, 8985.700823, 9044.654437, 29490....  \n",
       "4  [6540.401419, 6668.261153, 6475.719437, 19448....  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(miRNA_dataset)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772fa321",
   "metadata": {},
   "source": [
    "## Data preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "338bbaa3",
   "metadata": {},
   "source": [
    "creation of file with all miRNA base names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bd926b90",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.800534Z",
     "start_time": "2025-10-23T16:22:59.773802Z"
    }
   },
   "outputs": [],
   "source": [
    "genes_names = []\n",
    "#path = 'D:\\\\Universita\\\\2 anno magistrale\\\\Progetto BioInf\\\\miRNA_to_age\\\\datasets\\\\miRNA_seq\\\\0b64742b-b99d-4fc1-bfbf-bb7074f21a67\\\\9293a93f-b3d1-41ab-8c18-3deed4cc776d.mirbase21.mirnas.quantification.txt'\n",
    "path = '../datasets/miRNA_seq/0b64742b-b99d-4fc1-bfbf-bb7074f21a67/9293a93f-b3d1-41ab-8c18-3deed4cc776d.mirbase21.mirnas.quantification.txt'\n",
    "with open(path, 'r') as f: \n",
    "    lines = f.readlines()[1:]\n",
    "    for l in lines:\n",
    "        genes_names.append(l.split('\\t')[0])\n",
    "\n",
    "with open(os.path.join(DATA_PATH, 'genes_names.txt'), 'w') as f:\n",
    "    for name in genes_names:\n",
    "        f.write(name + '\\n') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec543277",
   "metadata": {},
   "source": [
    "extraction of miRNA_reads and adjustment into dataframe with each column being a gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2fa38438",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:40:31.312716Z",
     "start_time": "2025-10-23T18:40:31.262718Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3    hsa-let-7b   hsa-let-7c  \\\n",
      "Sample_0  12120.990742  12041.557881  12141.515730  17683.610218  2483.291559   \n",
      "Sample_1  20904.219040  21106.268560  20907.030164  45794.611568  6957.179909   \n",
      "Sample_2   4742.042712   4748.951756   4919.239371   4809.913909   546.220895   \n",
      "Sample_3   8992.901265   8985.700823   9044.654437  29490.757783  3047.586817   \n",
      "Sample_4   6540.401419   6668.261153   6475.719437  19448.969676  2930.996358   \n",
      "\n",
      "          hsa-let-7d   hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  hsa-let-7g  \\\n",
      "Sample_0  276.913390   860.889868   3728.010288   3782.511667  375.827595   \n",
      "Sample_1  471.917401  2279.821367   7539.785307   7636.417687  759.354805   \n",
      "Sample_2  298.714552   869.320309   1083.094260   1083.907089  495.419101   \n",
      "Sample_3  825.575608   577.835421   2360.394692   2396.171885  974.759753   \n",
      "Sample_4  310.623940   563.334943   1192.103985   1270.324057  559.574362   \n",
      "\n",
      "          ...  hsa-mir-941-5  hsa-mir-942  hsa-mir-943  hsa-mir-944  \\\n",
      "Sample_0  ...            0.0     3.478811          0.0     1.739406   \n",
      "Sample_1  ...            0.0     1.054171          0.0     1.405562   \n",
      "Sample_2  ...            0.0    13.411674          0.0     1.219243   \n",
      "Sample_3  ...            0.0    18.901159          0.0     0.900055   \n",
      "Sample_4  ...            0.0     1.504232          0.0     0.000000   \n",
      "\n",
      "          hsa-mir-95  hsa-mir-9500  hsa-mir-96  hsa-mir-98  hsa-mir-99a  \\\n",
      "Sample_0    1.159604           0.0   25.279363   33.860431   800.126629   \n",
      "Sample_1    4.919467           0.0    2.811124   52.708570  1664.888045   \n",
      "Sample_2    5.689801           0.0   98.352274   20.320718   293.837580   \n",
      "Sample_3    4.950303           0.0   19.801214   67.279124  1639.225485   \n",
      "Sample_4    8.273277           0.0   25.571947   30.836759   750.611846   \n",
      "\n",
      "            hsa-mir-99b  \n",
      "Sample_0   14239.470942  \n",
      "Sample_1   45043.338744  \n",
      "Sample_2  153860.753500  \n",
      "Sample_3   39918.797022  \n",
      "Sample_4   15080.679492  \n",
      "\n",
      "[5 rows x 1881 columns]\n",
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    }
   ],
   "source": [
    "def parse_array(x):\n",
    "    if isinstance(x, str):\n",
    "        x = x.strip(\"[]\")\n",
    "        return np.array([float(i) for i in x.split(\",\")])\n",
    "    \n",
    "    return np.array(x)\n",
    "n_cols = len(data[\"reads_per_million_miRNA_mapped\"][0].strip(\"[]\").split(\",\"))\n",
    "reads = data[\"reads_per_million_miRNA_mapped\"].apply(parse_array)\n",
    "genes_reads = pd.DataFrame(\n",
    "    np.stack(reads.values),  # Converte la Series di vettori in un array 2D\n",
    "    index=[f'Sample_{i}' for i in data.index]\n",
    ")\n",
    "f = open(os.path.join(DATA_PATH, 'genes_names.txt'), 'r')\n",
    "f_lines = f.readlines()\n",
    "genes_reads.columns = [i.strip() for i in f_lines]\n",
    "print(genes_reads.head())\n",
    "print(type(genes_reads))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f70f7e",
   "metadata": {},
   "source": [
    "### Extraction from web and manual dictionary of the corresponendce between miRNA base names and gene_id and gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a1ce1344",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:21.270091Z",
     "start_time": "2025-10-23T16:23:20.985272Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Gene stable ID  Gene name      miRBase ID\n",
      "0  ENSG00000283344  MIR1244-4  hsa-mir-1244-1\n",
      "1  ENSG00000283344  MIR1244-4  hsa-mir-1244-2\n",
      "2  ENSG00000283344  MIR1244-4  hsa-mir-1244-3\n",
      "3  ENSG00000283344  MIR1244-4  hsa-mir-1244-4\n",
      "4  ENSG00000292346    MIR6089  hsa-mir-6089-1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2170, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pybiomart import Server\n",
    "\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "mart = server['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = mart['hsapiens_gene_ensembl']\n",
    "\n",
    "# Query con filtro sui geni miRNA\n",
    "miRNA_mapping = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name', 'mirbase_id'],\n",
    "    filters={'biotype': 'miRNA'}  # filtra solo i miRNA\n",
    ")\n",
    "\n",
    "print(miRNA_mapping.head())\n",
    "miRNA_mapping.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b1d8d3161cf5e0a9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:23.906180Z",
     "start_time": "2025-10-23T16:23:23.677326Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene stable ID Gene name    miRBase ID\n",
      "2   ENSG00000252695   MIR2276  hsa-mir-2276\n",
      "5   ENSG00000263399   MIR3170  hsa-mir-3170\n",
      "7   ENSG00000207719    MIR623   hsa-mir-623\n",
      "9   ENSG00000263615   MIR4306  hsa-mir-4306\n",
      "12  ENSG00000265164   MIR2681  hsa-mir-2681\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1601, 3)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server = Server(host='http://grch37.ensembl.org')\n",
    "mart = server['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = mart['hsapiens_gene_ensembl']\n",
    "\n",
    "mapping_grch37 = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name', 'mirbase_id'],\n",
    "    filters={'biotype': 'miRNA'}\n",
    ")\n",
    "miRNA_mapping_grch37 = mapping_grch37[mapping_grch37['miRBase ID'].notnull()].drop_duplicates()\n",
    "\n",
    "print(miRNA_mapping_grch37.head())\n",
    "miRNA_mapping_grch37.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e2e188",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:26.053746Z",
     "start_time": "2025-10-23T16:23:25.998692Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature  start    end score strand frame  \\\n",
       "0    chr1  ENSEMBL        gene  17369  17436     .      -     .   \n",
       "1    chr1  ENSEMBL  transcript  17369  17436     .      -     .   \n",
       "2    chr1  ENSEMBL        exon  17369  17436     .      -     .   \n",
       "3    chr1  ENSEMBL        gene  30366  30503     .      +     .   \n",
       "4    chr1  ENSEMBL  transcript  30366  30503     .      +     .   \n",
       "\n",
       "             gene_id gene_type  ... exon_number            exon_id  \\\n",
       "0  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n",
       "1  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n",
       "2  ENSG00000278267.1     miRNA  ...         1.0  ENSE00003746039.1   \n",
       "3  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n",
       "4  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n",
       "\n",
       "  transcript_support_level havana_transcript     hgnc_id havana_gene  ont  \\\n",
       "0                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "1                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "2                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "3                      NaN               NaN  HGNC:35294         NaN  NaN   \n",
       "4                      NaN               NaN  HGNC:35294         NaN  NaN   \n",
       "\n",
       "  protein_id  ccdsid  artif_dupl  \n",
       "0        NaN     NaN         NaN  \n",
       "1        NaN     NaN         NaN  \n",
       "2        NaN     NaN         NaN  \n",
       "3        NaN     NaN         NaN  \n",
       "4        NaN     NaN         NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_miRNA = pd.read_csv(os.path.join(DATA_PATH, 'miRNA_gtf.csv'))\n",
    "gencode_miRNA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0c629d1c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:27:12.906840Z",
     "start_time": "2025-10-23T18:23:45.451456Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-1</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-2</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-3</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7c</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_id  gene_length\n",
       "0  hsa-let-7a-1           80\n",
       "1  hsa-let-7a-2           72\n",
       "2  hsa-let-7a-3           74\n",
       "3    hsa-let-7b           83\n",
       "4    hsa-let-7c           84"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manual_mapping = {\n",
    "    \"hsa-mir-3607\": \"MIR3607\",\n",
    "    \"hsa-mir-3653\": \"MIR3653\",\n",
    "    \"hsa-mir-3687-1\": \"MIR3687-1\",\n",
    "    \"hsa-mir-3687-2\": \"MIR3687-2\",\n",
    "    \"hsa-mir-6087\": \"MIR6087\",\n",
    "    \"hsa-mir-6723\": \"MIR6723\",\n",
    "    \"hsa-mir-6827\": \"MIR6827\",\n",
    "    \"hsa-mir-7641-1\": \"MIR7641-1\",\n",
    "    \"hsa-mir-7641-2\": \"MIR7641-2\",\n",
    "    \"hsa-mir-3656\": \"MIR3656\",\n",
    "    \"hsa-mir-4788\": \"MIR4788\"\n",
    "}\n",
    "\n",
    "gene_lengths = {}\n",
    "for name in genes_names:\n",
    "    gene_id_value = None\n",
    "    if name in manual_mapping:\n",
    "        gene_id_value = manual_mapping[name]\n",
    "    else:\n",
    "        map = miRNA_mapping[miRNA_mapping.astype(str).apply(lambda x: x.str.contains(name, case=False, na=False)).any(axis=1)]\n",
    "        if map.empty:\n",
    "            map = miRNA_mapping_grch37[miRNA_mapping_grch37.astype(str).apply(lambda x: x.str.contains(name, case=False, na=False)).any(axis=1)]\n",
    "        if map.empty:\n",
    "            print(f\"Gene not found for miRBase ID: {name}\")\n",
    "            continue            \n",
    "        gene_id_value = map['Gene name'].iloc[0] if not map['Gene name'].empty else None\n",
    "    \n",
    "    result = pd.DataFrame()\n",
    "    if gene_id_value:\n",
    "        result = gencode_miRNA[gencode_miRNA.astype(str).apply(lambda x: x.str.contains(gene_id_value, case=False, na=False)).any(axis=1)]\n",
    "    \n",
    "    if result.empty:\n",
    "        # Result is empty, trying with Gene stable ID\n",
    "        gene_stable_id = map['Gene stable ID'].iloc[0]\n",
    "        result = gencode_miRNA[gencode_miRNA.astype(str).apply(lambda x: x.str.contains(gene_stable_id, case=False, na=False)).any(axis=1)]\n",
    "        \n",
    "        if result.empty:\n",
    "            print(f\"Gene not found for miRBase ID: {name} using both Gene name and Gene stable ID\")\n",
    "            continue\n",
    "      \n",
    "    gene_lengths[name] = int(result['end'].values[0]) - int(result['start'].values[0]) + 1\n",
    "\n",
    "gene_lengths_df = pd.DataFrame([\n",
    "    {'gene_id': name, 'gene_length': length}\n",
    "    for name, length in gene_lengths.items()\n",
    "])\n",
    "gene_lengths_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acd9e60befa03895",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "The manual mapping was made consulting the miRBase database at https://www.mirbase.org/ ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "987d921a8285f28a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:39:29.766614Z",
     "start_time": "2025-10-23T18:39:29.755617Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1881, 2)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_lengths_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1ce50e5d65552ad0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:39:52.857787Z",
     "start_time": "2025-10-23T18:39:52.844787Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_id        0\n",
       "gene_length    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_lengths_df.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4051cfc8",
   "metadata": {},
   "source": [
    "creation of file with miRNA base names associated with their lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4e0eedc1dd3c8e74",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:39:09.816324Z",
     "start_time": "2025-10-23T18:39:09.777807Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "gene_lengths_df.to_csv(\"gene_lengths.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2039f2c",
   "metadata": {},
   "source": [
    "### TPM normalization application"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "bb6289de",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:40:15.296463Z",
     "start_time": "2025-10-23T18:40:13.151423Z"
    }
   },
   "outputs": [],
   "source": [
    "def tpm_norm(genes_reads, gene_lengths):\n",
    "    # Calcola il rapporto reads / lunghezza per ogni gene\n",
    "    reads_per_length = genes_reads.div(gene_lengths['gene_length'].values, axis=1)\n",
    "    \n",
    "    # Calcola la somma dei rapporti reads / lunghezza per ogni campione\n",
    "    scaling_factor = reads_per_length.sum(axis=1)\n",
    "    \n",
    "    # Calcola il TPM normalizzando e moltiplicando per 1e6\n",
    "    tpm = reads_per_length.div(scaling_factor, axis=0) * 1e6\n",
    "    \n",
    "    return tpm\n",
    "\n",
    "tmp = genes_reads.copy()\n",
    "normalized_reads = tpm_norm(tmp, gene_lengths_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b955782e",
   "metadata": {},
   "source": [
    "### TMM normalization application"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "a041c843",
   "metadata": {},
   "outputs": [
    {
     "ename": "NotImplementedError",
     "evalue": "\n    Conversion rules for `rpy2.robjects` appear to be missing. Those\n    rules are in a Python `contextvars.ContextVar`. This could be caused\n    by multithreading code not passing context to the thread.\n    Check rpy2's documentation about conversions.\n    ",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNotImplementedError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[115]\u001b[39m\u001b[32m, line 7\u001b[39m\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mrpy2\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mrobjects\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m pandas2ri\n\u001b[32m      5\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mrpy2\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mrobjects\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mpackages\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m importr\n\u001b[32m----> \u001b[39m\u001b[32m7\u001b[39m edgeR = \u001b[43mimportr\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43medgeR\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      9\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mtmm_normalization\u001b[39m(counts_df):\n\u001b[32m     10\u001b[39m     \u001b[38;5;66;03m# Converte il DataFrame pandas in un oggetto R\u001b[39;00m\n\u001b[32m     11\u001b[39m     counts_r = pandas2ri.py2rpy(counts_df)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\rpy2\\robjects\\packages.py:496\u001b[39m, in \u001b[36mimportr\u001b[39m\u001b[34m(name, lib_loc, robject_translations, signature_translation, suppress_messages, on_conflict, symbol_r2python, symbol_resolve, data)\u001b[39m\n\u001b[32m    494\u001b[39m pack: typing.Union[InstalledSTPackage, InstalledPackage]\n\u001b[32m    495\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m signature_translation:\n\u001b[32m--> \u001b[39m\u001b[32m496\u001b[39m     pack = \u001b[43mInstalledSTPackage\u001b[49m\u001b[43m(\u001b[49m\u001b[43menv\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    497\u001b[39m \u001b[43m                              \u001b[49m\u001b[43mtranslation\u001b[49m\u001b[43m=\u001b[49m\u001b[43mrobject_translations\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    498\u001b[39m \u001b[43m                              \u001b[49m\u001b[43mexported_names\u001b[49m\u001b[43m=\u001b[49m\u001b[43mexported_names\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    499\u001b[39m \u001b[43m                              \u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[43m=\u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    500\u001b[39m \u001b[43m                              \u001b[49m\u001b[43mversion\u001b[49m\u001b[43m=\u001b[49m\u001b[43mversion\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    501\u001b[39m \u001b[43m                              \u001b[49m\u001b[43msymbol_r2python\u001b[49m\u001b[43m=\u001b[49m\u001b[43msymbol_r2python\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    502\u001b[39m \u001b[43m                              \u001b[49m\u001b[43msymbol_resolve\u001b[49m\u001b[43m=\u001b[49m\u001b[43msymbol_resolve\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    503\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    504\u001b[39m     pack = InstalledPackage(env, name, translation=robject_translations,\n\u001b[32m    505\u001b[39m                             exported_names=exported_names,\n\u001b[32m    506\u001b[39m                             on_conflict=on_conflict,\n\u001b[32m    507\u001b[39m                             version=version,\n\u001b[32m    508\u001b[39m                             symbol_r2python=symbol_r2python,\n\u001b[32m    509\u001b[39m                             symbol_resolve=symbol_resolve)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\rpy2\\robjects\\packages.py:203\u001b[39m, in \u001b[36mPackage.__init__\u001b[39m\u001b[34m(self, env, name, translation, exported_names, on_conflict, version, symbol_r2python, symbol_resolve)\u001b[39m\n\u001b[32m    201\u001b[39m \u001b[38;5;28mself\u001b[39m._symbol_r2python = symbol_r2python\n\u001b[32m    202\u001b[39m \u001b[38;5;28mself\u001b[39m._symbol_resolve = symbol_resolve\n\u001b[32m--> \u001b[39m\u001b[32m203\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m__fill_rpy2r__\u001b[49m\u001b[43m(\u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[43m=\u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    204\u001b[39m \u001b[38;5;28mself\u001b[39m._exported_names = \u001b[38;5;28mself\u001b[39m._exported_names.difference(mynames)\n\u001b[32m    205\u001b[39m \u001b[38;5;28mself\u001b[39m.__version__ = version\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\rpy2\\robjects\\packages.py:330\u001b[39m, in \u001b[36mInstalledSTPackage.__fill_rpy2r__\u001b[39m\u001b[34m(self, on_conflict)\u001b[39m\n\u001b[32m    328\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m__fill_rpy2r__\u001b[39m(\u001b[38;5;28mself\u001b[39m, on_conflict=\u001b[33m'\u001b[39m\u001b[33mfail\u001b[39m\u001b[33m'\u001b[39m):\n\u001b[32m    329\u001b[39m     (\u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mSignatureTranslatedPackage\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m--> \u001b[39m\u001b[32m330\u001b[39m \u001b[43m     \u001b[49m\u001b[43m.\u001b[49m\u001b[43m__fill_rpy2r__\u001b[49m\u001b[43m(\u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[43m=\u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[32m    331\u001b[39m     \u001b[38;5;28;01mfor\u001b[39;00m name, robj \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m.\u001b[34m__dict__\u001b[39m.items():\n\u001b[32m    332\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(robj, rinterface.Sexp) \u001b[38;5;129;01mand\u001b[39;00m \\\n\u001b[32m    333\u001b[39m            robj.typeof == rinterface.RTYPES.CLOSXP:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\rpy2\\robjects\\packages.py:264\u001b[39m, in \u001b[36mPackage.__fill_rpy2r__\u001b[39m\u001b[34m(self, on_conflict)\u001b[39m\n\u001b[32m    262\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m rinterface.embedded.RRuntimeError \u001b[38;5;28;01mas\u001b[39;00m rre:\n\u001b[32m    263\u001b[39m     warn(\u001b[38;5;28mstr\u001b[39m(rre))\n\u001b[32m--> \u001b[39m\u001b[32m264\u001b[39m rpyobj = \u001b[43mcv\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrpy2py\u001b[49m\u001b[43m(\u001b[49m\u001b[43mriobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    265\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(rpyobj, \u001b[33m'\u001b[39m\u001b[33m__rname__\u001b[39m\u001b[33m'\u001b[39m):\n\u001b[32m    266\u001b[39m     rpyobj.__rname__ = rname\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\functools.py:912\u001b[39m, in \u001b[36msingledispatch.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kw)\u001b[39m\n\u001b[32m    908\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m args:\n\u001b[32m    909\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfuncname\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m requires at least \u001b[39m\u001b[33m'\u001b[39m\n\u001b[32m    910\u001b[39m                     \u001b[33m'\u001b[39m\u001b[33m1 positional argument\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m912\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdispatch\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m[\u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m.\u001b[49m\u001b[34;43m__class__\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mc:\\Users\\Matteo Bulgarelli\\anaconda3\\envs\\tirocinio-venv\\Lib\\site-packages\\rpy2\\robjects\\conversion.py:408\u001b[39m, in \u001b[36m_raise_missingconverter\u001b[39m\u001b[34m(obj)\u001b[39m\n\u001b[32m    401\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m_raise_missingconverter\u001b[39m(obj):\n\u001b[32m    402\u001b[39m     _missingconverter_msg = \u001b[33m\"\"\"\u001b[39m\n\u001b[32m    403\u001b[39m \u001b[33m    Conversion rules for `rpy2.robjects` appear to be missing. Those\u001b[39m\n\u001b[32m    404\u001b[39m \u001b[33m    rules are in a Python `contextvars.ContextVar`. This could be caused\u001b[39m\n\u001b[32m    405\u001b[39m \u001b[33m    by multithreading code not passing context to the thread.\u001b[39m\n\u001b[32m    406\u001b[39m \u001b[33m    Check rpy2\u001b[39m\u001b[33m'\u001b[39m\u001b[33ms documentation about conversions.\u001b[39m\n\u001b[32m    407\u001b[39m \u001b[33m    \u001b[39m\u001b[33m\"\"\"\u001b[39m\n\u001b[32m--> \u001b[39m\u001b[32m408\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m(_missingconverter_msg)\n",
      "\u001b[31mNotImplementedError\u001b[39m: \n    Conversion rules for `rpy2.robjects` appear to be missing. Those\n    rules are in a Python `contextvars.ContextVar`. This could be caused\n    by multithreading code not passing context to the thread.\n    Check rpy2's documentation about conversions.\n    "
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.environ['RPY2_CFFI_MODE'] = 'ABI'\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "edgeR = importr('edgeR')\n",
    "\n",
    "def tmm_normalization(counts_df):\n",
    "    # Converte il DataFrame pandas in un oggetto R\n",
    "    counts_r = pandas2ri.py2rpy(counts_df)\n",
    "    \n",
    "    # Crea un DGEList (oggetto edgeR)\n",
    "    dge_list = edgeR.DGEList(counts=counts_r)\n",
    "    \n",
    "    # Calcola i fattori di normalizzazione TMM\n",
    "    dge_list = edgeR.calcNormFactors(dge_list, method=\"TMM\")\n",
    "    \n",
    "    # Applica la normalizzazione\n",
    "    normalized_counts = edgeR.cpm(dge_list, normalized_lib_sizes=True)\n",
    "    \n",
    "    # Converte il risultato in un DataFrame pandas\n",
    "    normalized_df = pandas2ri.rpy2py(normalized_counts)\n",
    "    normalized_df.index = counts_df.index\n",
    "    normalized_df.columns = counts_df.columns\n",
    "    \n",
    "    return normalized_df\n",
    "\n",
    "normalized_reads = tmm_normalization(genes_reads)\n",
    "print(normalized_reads.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91ca679",
   "metadata": {},
   "source": [
    "### Creation of normalized file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "d180a663",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                case_id  age_at_initial_pathologic_diagnosis  \\\n",
      "0  378778d2-b331-4867-a93b-c64028c8b4c7                                 71.0   \n",
      "1  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda                                 53.0   \n",
      "2  b343bfe0-7c23-4c6a-8c84-9ee39db2ecda                                 53.0   \n",
      "3  3e775c99-ceda-4246-8d6f-0f58ca5097c8                                 59.0   \n",
      "4  abdc76db-f85e-4337-a57e-6d098789da03                                 55.0   \n",
      "\n",
      "                      reads_per_million_miRNA_mapped  \n",
      "0  [12493.837180157701, 13791.06770348918, 13529....  \n",
      "1  [21924.371432735934, 24595.868028313405, 23705...  \n",
      "2  [4556.89107219556, 5070.589282931563, 5110.453...  \n",
      "3  [9076.313794300138, 10076.718406113601, 9868.6...  \n",
      "4  [6345.182954300356, 7188.029244389861, 6791.81...  \n"
     ]
    }
   ],
   "source": [
    "final_merged_miRNA = pd.DataFrame({\n",
    "    'vector': normalized_reads.apply(lambda row: np.array(row.values), axis=1)\n",
    "})\n",
    "\n",
    "df = pd.read_csv(miRNA_dataset)\n",
    "df['reads_per_million_miRNA_mapped'] = final_merged_miRNA['vector'].values\n",
    "print(df.head())\n",
    "df.to_csv(os.path.join(DATA_PATH, 'final_merged_miRNA_normalized.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tirocinio-venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
