{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bda669cc",
   "metadata": {},
   "source": [
    "# DATA NORMALIZATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2d88d76",
   "metadata": {},
   "source": [
    "## Theory"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b436641",
   "metadata": {},
   "source": [
    "Various factors affect transcript quantification in RNA-seq data, such as sequencing depth, transcript length, and sample-to-sample and batch-to-batch variability. Normalization methods exist to minimize these variables and ensure reliable transcriptomic data. Sequencing technologies introduce technical variability. Therefore, raw transcriptomic data must be adjusted to account for these technical factors. \n",
    "\n",
    "It is essential to choose the correct RNA-seq normalization method for the dataset and there are three main RNA-seq normalization stages to consider:\n",
    " 1. **Within sample**: Within sample normalization is required to compare the expression of genes within an individual sample. It can adjust data for two primary technical variables: transcript length and sequencing depth. Longer genes often have more mapped reads than shorter genes at the same expression level. Therefore, their expression level can only be accurately compared within a sample after normalization. Furthermore, the number of sequencing reads per sample may vary. This can also be corrected by within sample normalization. Within sample normalization is not sufficient to compare gene expression between samples. For this, between sample RNA-seq normalization methods are required. Within sample normalization most common techniques are:\n",
    "    - **CPM**: Counts per million (CPM) mapped reads are the number of raw reads mapped to a transcript, scaled by the number of sequencing reads in your sample, multiplied by a million. It normalizes RNA-seq data for sequencing depth but not gene length. Therefore, although it is a within sample normalization approach, CPM normalization is unsuitable for within sample comparisons of gene expression. Between sample comparisons can be made when CPM is used alongside ‘within a dataset’ normalization methods.\n",
    "    - **FPKM/RPKM**: FPKM (fragments per kilobase of transcript per million fragments mapped) for paired-end data and RPKM (reads per kilobase of transcript per million reads mapped) for single-end data correct for variations in library size and gene length. One issue with FPKM/RPKM units is that the expression of a gene in one sample will appear different from its expression in another sample, even when its true expression level is the same. This is because it depends on the relative abundance of a transcript among a population of sequenced transcripts. FPKM/RPKM units best compare gene expression within a single sample\n",
    "    - **TPM**: Transcripts per million (TPM) represents the relative number of transcripts you would detect for a gene if you had sequenced one million full-length transcripts. It is calculated by dividing the number of reads mapped to a transcript by the transcript length. This value is then divided by the sum of mapped reads to all transcripts after normalization for transcript length. It is then multiplied by one million to allow easier further analyses. It normalizes RNA-seq data for sequencing depth and transcript length. TPM and FPKM/RPKM are closely related, however, in contrast to FPKM/RPKM, there is limited variation in values between samples as the sum of all TPMs in each sample is the same. TPM can be used for within sample comparisons but requires ‘within a dataset’ normalization for between sample comparisons.\n",
    "\n",
    " 2. **Within a dataset (between samples)**: Samples within a dataset can be simultaneously normalized as a complete set to adjust for different technical variations such as sequencing depth. RNA-seq is a relative, not an absolute, measure of transcript abundance. This means that the transcript population as a whole affects relative levels of transcripts. This creates biases for gene expression analyses, and these are minimized by between sample RNA-seq normalization methods. Within a dataset normalization mosto common techniques are:\n",
    "    - **Quantile**: The quantile method aims to make the distribution of gene expression levels the same for each sample in a dataset. It assumes that the global differences in distributions between samples are all due to technical variation. Any remaining differences are likely actual biological effects. For each sample, genes are ranked based on their expression level. An average value is calculated across all samples for genes of the same rank. This average value then replaces the original value of all genes in that rank. These genes are then placed in their original order.\n",
    "    - **TMM**: TMM (trimmed mean of M-values) also assumes that most genes are not differentially expressed between samples. If many genes are uniquely or highly expressed in one experimental condition, it will affect the accurate quantification of the remaining genes. To adjust for this possibility, TMM calculates scaling factors to adjust library sizes for the normalization of samples within a dataset. To do this, one sample is chosen as a reference sample. The fold changes and absolute expression levels of other samples within the dataset are then calculated relative to the reference sample. Next, the genes in the data set are ‘trimmed’ to remove differentially expressed genes using these two values. The trimmed mean of the fold changes is then found for each sample. Finally, read counts are scaled by this trimmed mean and the total count of their sample.\n",
    "\n",
    " 3. **Across datasets**: Researchers often integrate RNA-seq data from multiple independent studies. These datasets are usually sequenced at different times, with varying methods across multiple facilities, and contain other experimental variables. This results in a batch effect. The batch effect is often responsible for the greatest source of differential expression when data is combined. It can mask any true biological differences and lead to incorrect conclusions. RNA-seq normalization across datasets can correct for known variables across batches, such as the sequencing center and date of sequencing, as well as unknown variables.\n",
    "\n",
    "source : https://bigomics.ch/blog/why-how-normalize-rna-seq-data/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536b5bb5",
   "metadata": {},
   "source": [
    "**PERSONAL CONSIDERATIONS**\n",
    "\n",
    "Since we want to have a model capable of inferring the survivability of a patient given a feature vector, we thought that the model would need, to behave at best, data normalized across the dataset allowing intra-dataset comparisons in order to learn when a feature vector corresponds to higher or lower survivability. Given this preposition we first normalize miRNA read values with the TMM method to allow genetic intra-dataset comparisons, and then normalize all values of the feature vectors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c25e1516",
   "metadata": {},
   "source": [
    "## Init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d0b19a22",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.257776Z",
     "start_time": "2025-10-23T16:22:58.159659Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f47107ac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.288372Z",
     "start_time": "2025-10-23T16:22:59.262349Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "d:\\Universita\\2 anno magistrale\\Progetto BioInf\\miRNA_to_age\n"
     ]
    }
   ],
   "source": [
    "base = os.path.basename(os.getcwd())\n",
    "list = os.getcwd().split(os.sep) \n",
    "list.pop(list.index(base))\n",
    "ROOT = '\\\\'.join(list)\n",
    "print(ROOT)\n",
    "DATA_PATH = os.path.join(ROOT, 'datasets\\\\preprocessed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a9377bf2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:22:59.769799Z",
     "start_time": "2025-10-23T16:22:59.326271Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['days_to_death', 'pathologic_stage',\n",
      "       'age_at_initial_pathologic_diagnosis', 'days_to_last_followup', 'Death',\n",
      "       'case_id', 'miRNA_ID', 'read_count', 'reads_per_million_miRNA_mapped'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "raw_data = pd.read_csv(os.path.join(DATA_PATH, 'clinical_miRNA(RC_RPM).csv'))\n",
    "print(raw_data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1ad104a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from sklearn.preprocessing import OneHotEncoder\n",
    "\n",
    "# print(data.shape)\n",
    "\n",
    "# tmp = data.drop(columns=['radiation_therapy', 'case_id'])\n",
    "# print(tmp.shape)\n",
    "# tmp = tmp.dropna(subset=['pathologic_stage'])\n",
    "# print(tmp.shape)\n",
    "\n",
    "# tmp.to_csv(os.path.join(DATA_PATH, 'clinical_miRNA.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772fa321",
   "metadata": {},
   "source": [
    "## Data preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dab3a0cd",
   "metadata": {},
   "source": [
    "Removed useless columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "519389f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(767, 9)\n",
      "(767, 8)\n"
     ]
    }
   ],
   "source": [
    "print(raw_data.shape)\n",
    "raw_data.drop(columns=['case_id'], inplace=True)\n",
    "print(raw_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7e6aa270",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_array(x):\n",
    "    if isinstance(x, str):\n",
    "        x = x.strip(\"[]\")\n",
    "        return np.array(eval(x, {'np':np}))\n",
    "    \n",
    "    return np.array(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6a9343",
   "metadata": {},
   "source": [
    "One-hot encoding of pathological stage and deletion of columns with less than 20 samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8da3b5ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(767, 8)\n",
      "(767, 12)\n",
      "(767, 6)\n",
      "Index(['days_to_death', 'age_at_initial_pathologic_diagnosis',\n",
      "       'days_to_last_followup', 'Death', 'miRNA_ID', 'read_count',\n",
      "       'reads_per_million_miRNA_mapped', 'pathologic_stage_Stage I',\n",
      "       'pathologic_stage_Stage IA', 'pathologic_stage_Stage IIA',\n",
      "       'pathologic_stage_Stage IIB', 'pathologic_stage_Stage IIIA',\n",
      "       'pathologic_stage_Stage IIIC'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "from sklearn.preprocessing import OneHotEncoder\n",
    "\n",
    "print(raw_data.shape)\n",
    "\n",
    "pathologic_stage = raw_data[['pathologic_stage']]\n",
    "\n",
    "encoder = OneHotEncoder()\n",
    "encoded_stages = encoder.fit_transform(pathologic_stage).toarray()\n",
    "encoded_columns = encoder.get_feature_names_out(['pathologic_stage'])\n",
    "encoded_df = pd.DataFrame(encoded_stages, columns=encoded_columns, index=pathologic_stage.index)\n",
    "\n",
    "print(encoded_df.shape)\n",
    "index = encoded_df.sum().index\n",
    "encoded_df.drop(columns=[i for i in index if encoded_df.sum()[i]<20], inplace=True)\n",
    "print(encoded_df.shape)\n",
    "\n",
    "raw_data = pd.concat([raw_data, encoded_df], axis=1)\n",
    "raw_data = raw_data.drop(columns=['pathologic_stage'])\n",
    "print(raw_data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c8b72ec",
   "metadata": {},
   "source": [
    "Removal of age outliers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "2213fd8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(767, 13)\n",
      "(746, 13)\n"
     ]
    }
   ],
   "source": [
    "print(raw_data.shape)\n",
    "\n",
    "ages_distrib = raw_data['age_at_initial_pathologic_diagnosis'].value_counts()\n",
    "ages_to_del = ages_distrib[ages_distrib < 5].index.tolist()\n",
    "mask = raw_data['age_at_initial_pathologic_diagnosis'].isin(ages_to_del)\n",
    "indexes = raw_data[mask].index.to_list()\n",
    "raw_data.drop(indexes, inplace=True)\n",
    "print(raw_data.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec543277",
   "metadata": {},
   "source": [
    "extraction of miRNA_reads and adjustment into dataframe with each column being a gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fa38438",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:40:31.312716Z",
     "start_time": "2025-10-23T18:40:31.262718Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3    hsa-let-7b   hsa-let-7c  \\\n",
      "Sample_0   7988.580442   8018.197889   8100.662153  32345.736656  3366.516489   \n",
      "Sample_1  11857.268807  11728.983540  11841.861998  13907.317668  4529.287401   \n",
      "Sample_2   7918.388430   8062.278630   7963.400236  10422.815821  1934.769776   \n",
      "Sample_3  14251.643715  14324.037522  14391.970087  13755.431817  1047.783766   \n",
      "Sample_4   4818.534597   4785.049036   4915.228857   9827.447776  1689.327738   \n",
      "\n",
      "          hsa-let-7d   hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  hsa-let-7g  \\\n",
      "Sample_0  300.820345  2176.011262   2434.438006   2446.052691  184.673494   \n",
      "Sample_1  593.948205   996.411786   8635.673617   8693.842181  608.411740   \n",
      "Sample_2  340.909091   515.053129   2789.256198   2791.469894  323.199528   \n",
      "Sample_3  330.334770  3843.563673   6485.836263   6573.033286  412.867766   \n",
      "Sample_4  667.830008  4157.853192   1991.450271   2011.014868  780.702685   \n",
      "\n",
      "          ...  hsa-mir-941-5  hsa-mir-942  hsa-mir-943  hsa-mir-944  \\\n",
      "Sample_0  ...            0.0     2.322937          0.0     0.000000   \n",
      "Sample_1  ...            0.0     2.829822          0.0     1.257699   \n",
      "Sample_2  ...            0.0     6.641086          0.0     0.000000   \n",
      "Sample_3  ...            0.0     3.852892          0.0     0.608351   \n",
      "Sample_4  ...            0.0     4.138665          0.0     0.000000   \n",
      "\n",
      "          hsa-mir-95  hsa-mir-9500  hsa-mir-96  hsa-mir-98  hsa-mir-99a  \\\n",
      "Sample_0    1.161469           0.0    1.742203   78.399125   487.236042   \n",
      "Sample_1    0.943274           0.0   15.406809   78.291743   785.747255   \n",
      "Sample_2    1.475797           0.0   11.806375   24.350649  1133.412043   \n",
      "Sample_3    2.230622           0.0   37.109431  120.250779   317.762176   \n",
      "Sample_4    4.514907           0.0   66.971122   88.040688   480.837606   \n",
      "\n",
      "            hsa-mir-99b  \n",
      "Sample_0   59124.554722  \n",
      "Sample_1   16872.027901  \n",
      "Sample_2   28957.349469  \n",
      "Sample_3   52594.811047  \n",
      "Sample_4  164538.640456  \n",
      "\n",
      "[5 rows x 1881 columns]\n",
      "(746, 1881)\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'genes_reads(RPM).csv')):\n",
    "    genes_reads = pd.read_csv(os.path.join(DATA_PATH, 'genes_reads(RPM).csv'), index_col=0)\n",
    "    print(genes_reads.head())\n",
    "\n",
    "else:\n",
    "    n_cols = len(raw_data[\"reads_per_million_miRNA_mapped\"][0].strip(\"[]\").split(\",\"))\n",
    "    reads = raw_data[\"reads_per_million_miRNA_mapped\"].apply(parse_array)\n",
    "\n",
    "    genes_reads = pd.DataFrame(\n",
    "        np.stack(reads.values),  # Converte la Series di vettori in un array 2D\n",
    "        index=[f'Sample_{i}' for i in raw_data.index],\n",
    "        columns = np.array(eval(raw_data['miRNA_ID'][0].strip('[]'), {'np':str}))\n",
    "    )\n",
    "\n",
    "    genes_reads.to_csv(os.path.join(DATA_PATH, 'genes_reads(RPM).csv'))\n",
    "    print(genes_reads.head())\n",
    "\n",
    "print(genes_reads.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f70f7e",
   "metadata": {},
   "source": [
    "### Extraction from web and manual dictionary of the corresponendce between miRNA base names and gene_id and gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf423c59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              gene_length\n",
      "gene_id                  \n",
      "hsa-let-7a-1           80\n",
      "hsa-let-7a-2           72\n",
      "hsa-let-7a-3           74\n",
      "hsa-let-7b             83\n",
      "hsa-let-7c             84\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'gene_lengths.csv')):\n",
    "    gene_lengths_df = pd.read_csv(os.path.join(DATA_PATH, 'gene_lengths.csv'), sep=\"\\t\", index_col=0)\n",
    "    print(gene_lengths_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a1ce1344",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:21.270091Z",
     "start_time": "2025-10-23T16:23:20.985272Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Gene stable ID  Gene name      miRBase ID\n",
      "0  ENSG00000283344  MIR1244-4  hsa-mir-1244-1\n",
      "1  ENSG00000283344  MIR1244-4  hsa-mir-1244-2\n",
      "2  ENSG00000283344  MIR1244-4  hsa-mir-1244-3\n",
      "3  ENSG00000283344  MIR1244-4  hsa-mir-1244-4\n",
      "4  ENSG00000292346    MIR6089  hsa-mir-6089-1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2170, 3)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pybiomart import Server\n",
    "\n",
    "server = Server(host='http://www.ensembl.org')\n",
    "mart = server['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = mart['hsapiens_gene_ensembl']\n",
    "\n",
    "# Query con filtro sui geni miRNA\n",
    "miRNA_mapping = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name', 'mirbase_id'],\n",
    "    filters={'biotype': 'miRNA'}  # filtra solo i miRNA\n",
    ")\n",
    "\n",
    "print(miRNA_mapping.head())\n",
    "miRNA_mapping.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b1d8d3161cf5e0a9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:23.906180Z",
     "start_time": "2025-10-23T16:23:23.677326Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene stable ID Gene name    miRBase ID\n",
      "2   ENSG00000252695   MIR2276  hsa-mir-2276\n",
      "5   ENSG00000263399   MIR3170  hsa-mir-3170\n",
      "7   ENSG00000207719    MIR623   hsa-mir-623\n",
      "9   ENSG00000263615   MIR4306  hsa-mir-4306\n",
      "12  ENSG00000265164   MIR2681  hsa-mir-2681\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1601, 3)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server = Server(host='http://grch37.ensembl.org')\n",
    "mart = server['ENSEMBL_MART_ENSEMBL']\n",
    "dataset = mart['hsapiens_gene_ensembl']\n",
    "\n",
    "mapping_grch37 = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name', 'mirbase_id'],\n",
    "    filters={'biotype': 'miRNA'}\n",
    ")\n",
    "miRNA_mapping_grch37 = mapping_grch37[mapping_grch37['miRBase ID'].notnull()].drop_duplicates()\n",
    "\n",
    "print(miRNA_mapping_grch37.head())\n",
    "miRNA_mapping_grch37.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e2e188",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T16:23:26.053746Z",
     "start_time": "2025-10-23T16:23:25.998692Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSE00003746039.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname   source     feature  start    end score strand frame  \\\n",
       "0    chr1  ENSEMBL        gene  17369  17436     .      -     .   \n",
       "1    chr1  ENSEMBL  transcript  17369  17436     .      -     .   \n",
       "2    chr1  ENSEMBL        exon  17369  17436     .      -     .   \n",
       "3    chr1  ENSEMBL        gene  30366  30503     .      +     .   \n",
       "4    chr1  ENSEMBL  transcript  30366  30503     .      +     .   \n",
       "\n",
       "             gene_id gene_type  ... exon_number            exon_id  \\\n",
       "0  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n",
       "1  ENSG00000278267.1     miRNA  ...         NaN                NaN   \n",
       "2  ENSG00000278267.1     miRNA  ...         1.0  ENSE00003746039.1   \n",
       "3  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n",
       "4  ENSG00000284332.1     miRNA  ...         NaN                NaN   \n",
       "\n",
       "  transcript_support_level havana_transcript     hgnc_id havana_gene  ont  \\\n",
       "0                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "1                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "2                      NaN               NaN  HGNC:50039         NaN  NaN   \n",
       "3                      NaN               NaN  HGNC:35294         NaN  NaN   \n",
       "4                      NaN               NaN  HGNC:35294         NaN  NaN   \n",
       "\n",
       "  protein_id  ccdsid  artif_dupl  \n",
       "0        NaN     NaN         NaN  \n",
       "1        NaN     NaN         NaN  \n",
       "2        NaN     NaN         NaN  \n",
       "3        NaN     NaN         NaN  \n",
       "4        NaN     NaN         NaN  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_miRNA = pd.read_csv(os.path.join(DATA_PATH, 'miRNA_gtf.csv'))\n",
    "gencode_miRNA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c629d1c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-10-23T18:27:12.906840Z",
     "start_time": "2025-10-23T18:23:45.451456Z"
    }
   },
   "outputs": [],
   "source": [
    "manual_mapping = {\n",
    "    \"hsa-mir-3607\": \"MIR3607\",\n",
    "    \"hsa-mir-3653\": \"MIR3653\",\n",
    "    \"hsa-mir-3687-1\": \"MIR3687-1\",\n",
    "    \"hsa-mir-3687-2\": \"MIR3687-2\",\n",
    "    \"hsa-mir-6087\": \"MIR6087\",\n",
    "    \"hsa-mir-6723\": \"MIR6723\",\n",
    "    \"hsa-mir-6827\": \"MIR6827\",\n",
    "    \"hsa-mir-7641-1\": \"MIR7641-1\",\n",
    "    \"hsa-mir-7641-2\": \"MIR7641-2\",\n",
    "    \"hsa-mir-3656\": \"MIR3656\",\n",
    "    \"hsa-mir-4788\": \"MIR4788\"\n",
    "}\n",
    "\n",
    "gene_lengths = {}\n",
    "for name in genes_names:\n",
    "    gene_id_value = None\n",
    "    if name in manual_mapping:\n",
    "        gene_id_value = manual_mapping[name]\n",
    "    else:\n",
    "        map = miRNA_mapping[miRNA_mapping.astype(str).apply(lambda x: x.str.contains(name, case=False, na=False)).any(axis=1)]\n",
    "        if map.empty:\n",
    "            map = miRNA_mapping_grch37[miRNA_mapping_grch37.astype(str).apply(lambda x: x.str.contains(name, case=False, na=False)).any(axis=1)]\n",
    "        if map.empty:\n",
    "            print(f\"Gene not found for miRBase ID: {name}\")\n",
    "            continue            \n",
    "        gene_id_value = map['Gene name'].iloc[0] if not map['Gene name'].empty else None\n",
    "    \n",
    "    result = pd.DataFrame()\n",
    "    if gene_id_value:\n",
    "        result = gencode_miRNA[gencode_miRNA.astype(str).apply(lambda x: x.str.contains(gene_id_value, case=False, na=False)).any(axis=1)]\n",
    "    \n",
    "    if result.empty:\n",
    "        # Result is empty, trying with Gene stable ID\n",
    "        gene_stable_id = map['Gene stable ID'].iloc[0]\n",
    "        result = gencode_miRNA[gencode_miRNA.astype(str).apply(lambda x: x.str.contains(gene_stable_id, case=False, na=False)).any(axis=1)]\n",
    "        \n",
    "        if result.empty:\n",
    "            print(f\"Gene not found for miRBase ID: {name} using both Gene name and Gene stable ID\")\n",
    "            continue\n",
    "      \n",
    "    gene_lengths[name] = int(result['end'].values[0]) - int(result['start'].values[0]) + 1\n",
    "\n",
    "gene_lengths_df = pd.DataFrame([\n",
    "    {'gene_id': name, 'gene_length': length}\n",
    "    for name, length in gene_lengths.items()\n",
    "])\n",
    "\n",
    "print(gene_lengths_df.head())\n",
    "print(gene_lengths_df.shape)\n",
    "print(gene_lengths_df.isna().sum())\n",
    "gene_lengths_df.to_csv(\"gene_lengths.csv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acd9e60befa03895",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "The manual mapping was made consulting the miRBase database at https://www.mirbase.org/ ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ec42e51",
   "metadata": {},
   "source": [
    "### Quantile and Log2 normalization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e0b71cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3    hsa-let-7b   hsa-let-7c\n",
      "Sample_0   7988.580442   8018.197889   8100.662153  32345.736656  3366.516489\n",
      "Sample_1  11857.268807  11728.983540  11841.861998  13907.317668  4529.287401\n",
      "Sample_2   7918.388430   8062.278630   7963.400236  10422.815821  1934.769776\n",
      "Sample_3  14251.643715  14324.037522  14391.970087  13755.431817  1047.783766\n",
      "Sample_4   4818.534597   4785.049036   4915.228857   9827.447776  1689.327738\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Matteo Bulgarelli\\AppData\\Local\\Temp\\ipykernel_1540\\2233587725.py:6: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  reads_logged = genes_reads.applymap(lambda x: np.log2(float(x) + 1e-3))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c\n",
      "Sample_0     12.963724     12.969062     12.983824   14.981288   11.717041\n",
      "Sample_1     13.533484     13.517790     13.531608   13.763557   12.145069\n",
      "Sample_2     12.950991     12.976972     12.959169   13.347458   10.917947\n",
      "Sample_3     13.798841     13.806151     13.812977   13.747714   10.033127\n",
      "Sample_4     12.234379     12.224318     12.263043   13.262601   10.722234\n",
      "          hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c\n",
      "Sample_0     -6.647062     -6.631267     -6.633072   -5.583013   -4.765750\n",
      "Sample_1     -5.327739     -5.355312     -5.338547   -7.191785   -3.282310\n",
      "Sample_2     -6.697892     -6.617413     -6.689258   -7.578081   -6.268020\n",
      "Sample_3     -4.631995     -4.623910     -4.623910   -7.207690   -7.210804\n",
      "Sample_4     -7.639684     -7.642413     -7.639684   -7.619390   -6.622316\n"
     ]
    }
   ],
   "source": [
    "import qnorm\n",
    "\n",
    "print(genes_reads.iloc[:5, :5])\n",
    "\n",
    "# Log2 transform necessary for 0 values before quantile normalization\n",
    "reads_logged = genes_reads.applymap(lambda x: np.log2(float(x) + 1e-3))\n",
    "print(reads_logged.iloc[:5, :5])\n",
    "\n",
    "# Quantile normalization\n",
    "quant_norm = qnorm.quantile_normalize(reads_logged)\n",
    "print(quant_norm.iloc[:5, :5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77a8070",
   "metadata": {},
   "source": [
    "Delete reads columns with variance under the 50th percentile of the value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b0eb8a83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(746, 1881)\n",
      "Columns to drop:940\n",
      "(746, 941)\n"
     ]
    }
   ],
   "source": [
    "print(reads_logged.shape)\n",
    "zero_var_cols = reads_logged.var()[reads_logged.var()<reads_logged.var().describe().loc['50%']].index\n",
    "print(f\"Columns to drop:{len(zero_var_cols)}\")\n",
    "reads_logged.drop(columns=[i for i in reads_logged.columns if i in zero_var_cols], inplace=True)\n",
    "print(reads_logged.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "bda576b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(746, 1881)\n",
      "Columns to drop:938\n",
      "(746, 943)\n"
     ]
    }
   ],
   "source": [
    "print(quant_norm.shape)\n",
    "quant_norm_cols = quant_norm.var()[quant_norm.var()<quant_norm.var().describe().loc['50%']].index\n",
    "print(f\"Columns to drop:{len(quant_norm_cols)}\")\n",
    "quant_norm.drop(columns=[i for i in quant_norm.columns if i in quant_norm_cols], inplace=True)\n",
    "print(quant_norm.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91ca679",
   "metadata": {},
   "source": [
    "### Creation of normalized file with just log2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "d180a663",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating new dataset\n",
      "   days_to_death  age_at_initial_pathologic_diagnosis  days_to_last_followup  \\\n",
      "0           -1.0                                   71                 1918.0   \n",
      "1           -1.0                                   53                 1309.0   \n",
      "2           -1.0                                   55                    0.0   \n",
      "3           -1.0                                   64                  212.0   \n",
      "4         2763.0                                   46                 2763.0   \n",
      "\n",
      "   Death  pathologic_stage_Stage I  pathologic_stage_Stage IA  \\\n",
      "0      0                       0.0                        0.0   \n",
      "1      0                       1.0                        0.0   \n",
      "2      0                       0.0                        0.0   \n",
      "3      0                       0.0                        0.0   \n",
      "4      1                       1.0                        0.0   \n",
      "\n",
      "   pathologic_stage_Stage IIA  pathologic_stage_Stage IIB  \\\n",
      "0                         1.0                         0.0   \n",
      "1                         0.0                         0.0   \n",
      "2                         0.0                         1.0   \n",
      "3                         1.0                         0.0   \n",
      "4                         0.0                         0.0   \n",
      "\n",
      "   pathologic_stage_Stage IIIA  pathologic_stage_Stage IIIC  ...  hsa-mir-934  \\\n",
      "0                          0.0                          0.0  ...    -9.965784   \n",
      "1                          0.0                          0.0  ...     4.705881   \n",
      "2                          0.0                          0.0  ...     1.883813   \n",
      "3                          0.0                          0.0  ...    -0.714655   \n",
      "4                          0.0                          0.0  ...    -0.408349   \n",
      "\n",
      "   hsa-mir-935  hsa-mir-937  hsa-mir-938  hsa-mir-939  hsa-mir-940  \\\n",
      "0     3.970930     2.538127    -9.965784    -0.781568     3.739620   \n",
      "1    -0.666922     1.331360    -9.965784    -0.666922     2.500966   \n",
      "2     0.562472     0.562472    -9.965784     0.562472     2.731636   \n",
      "3     3.426069     2.983599    -9.965784     1.283569     3.531029   \n",
      "4     3.049340     0.912428    -9.965784    -1.406438     3.233743   \n",
      "\n",
      "   hsa-mir-942  hsa-mir-943  hsa-mir-944  hsa-mir-95  \n",
      "0     1.216571    -9.965784    -9.965784    0.217192  \n",
      "1     1.501221    -9.965784     0.331933   -0.082723  \n",
      "2     2.731636    -9.965784    -9.965784    0.562472  \n",
      "3     1.946316    -9.965784    -0.714655    1.158093  \n",
      "4     2.049514    -9.965784    -9.965784    2.175016  \n",
      "\n",
      "[5 rows x 951 columns]\n",
      "(746, 951)\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'clinical_miRNA_normalized_log.csv')):\n",
    "    print(\"Loading existing dataset\")\n",
    "    dataset = pd.read_csv(os.path.join(DATA_PATH, 'clinical_miRNA_normalized_log.csv'))    \n",
    "\n",
    "else:\n",
    "    print(\"Creating new dataset\")\n",
    "    dataset = raw_data.copy()\n",
    "    dataset.drop(columns=['read_count', 'reads_per_million_miRNA_mapped', 'miRNA_ID'], inplace=True)\n",
    "    \n",
    "    reads_logged.index = dataset.index\n",
    "    dataset = pd.concat([dataset, reads_logged], axis=1)\n",
    "    dataset.to_csv(os.path.join(DATA_PATH, 'clinical_miRNA_normalized_log.csv'), index=False)\n",
    "\n",
    "print(dataset.head())\n",
    "print(dataset.shape)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62fbf73b",
   "metadata": {},
   "source": [
    "### Creation of normalized file with also quantile norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "4413a977",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating new dataset\n",
      "   days_to_death  age_at_initial_pathologic_diagnosis  days_to_last_followup  \\\n",
      "0           -1.0                                   71                 1918.0   \n",
      "1           -1.0                                   53                 1309.0   \n",
      "2           -1.0                                   55                    0.0   \n",
      "3           -1.0                                   64                  212.0   \n",
      "4         2763.0                                   46                 2763.0   \n",
      "\n",
      "   Death  pathologic_stage_Stage I  pathologic_stage_Stage IA  \\\n",
      "0      0                       0.0                        0.0   \n",
      "1      0                       1.0                        0.0   \n",
      "2      0                       0.0                        0.0   \n",
      "3      0                       0.0                        0.0   \n",
      "4      1                       1.0                        0.0   \n",
      "\n",
      "   pathologic_stage_Stage IIA  pathologic_stage_Stage IIB  \\\n",
      "0                         1.0                         0.0   \n",
      "1                         0.0                         0.0   \n",
      "2                         0.0                         1.0   \n",
      "3                         1.0                         0.0   \n",
      "4                         0.0                         0.0   \n",
      "\n",
      "   pathologic_stage_Stage IIIA  pathologic_stage_Stage IIIC  ...  hsa-mir-939  \\\n",
      "0                          0.0                          0.0  ...    -7.099648   \n",
      "1                          0.0                          0.0  ...    -7.054477   \n",
      "2                          0.0                          0.0  ...    -5.942137   \n",
      "3                          0.0                          0.0  ...    -4.901064   \n",
      "4                          0.0                          0.0  ...    -7.298907   \n",
      "\n",
      "   hsa-mir-940  hsa-mir-942  hsa-mir-943  hsa-mir-944  hsa-mir-95  hsa-mir-96  \\\n",
      "0    -4.259288    -7.186019    -6.504231    -7.547096   -7.191785   -8.244224   \n",
      "1    -6.044533    -6.950198    -6.504231    -5.862583   -7.324579   -7.337426   \n",
      "2    -5.770512    -5.614530    -6.504231    -7.547096   -7.040831   -7.543822   \n",
      "3    -4.671122    -6.552662    -6.504231    -6.664567   -6.627685   -6.261447   \n",
      "4    -5.143548    -6.467418    -6.504231    -7.547096   -5.486871   -4.723142   \n",
      "\n",
      "   hsa-mir-98  hsa-mir-99a  hsa-mir-99b  \n",
      "0   -5.704588    -6.561818    -5.114681  \n",
      "1   -5.713534    -5.422679    -7.657616  \n",
      "2   -7.886832    -4.239332    -7.046339  \n",
      "3   -3.840576    -7.099648    -5.536121  \n",
      "4   -5.210531    -6.595688    -1.570993  \n",
      "\n",
      "[5 rows x 953 columns]\n",
      "(746, 953)\n"
     ]
    }
   ],
   "source": [
    "if os.path.exists(os.path.join(DATA_PATH, 'clinical_miRNA_normalized_quant.csv')):\n",
    "    print(\"Loading existing dataset\")\n",
    "    dataset = pd.read_csv(os.path.join(DATA_PATH, 'clinical_miRNA_normalized_quant.csv'))\n",
    "\n",
    "else:\n",
    "    print(\"Creating new dataset\")\n",
    "    dataset = raw_data.copy()\n",
    "    dataset.drop(columns=['read_count', 'reads_per_million_miRNA_mapped', 'miRNA_ID'], inplace=True)\n",
    "    \n",
    "    quant_norm.index = dataset.index\n",
    "    dataset = pd.concat([dataset, quant_norm], axis=1)\n",
    "    dataset.to_csv(os.path.join(DATA_PATH, 'clinical_miRNA_normalized_quant.csv'), index=False)\n",
    "\n",
    "print(dataset.head())\n",
    "print(dataset.shape)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tirocinio-venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
